---
title: "brain_damage"
output: html_document
date: "2024-12-07"
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library (tidyverse)
library (readxl)
library(lubridate)
library(tidycmprsk)
library(ggsurvfit)
library(anytime)
library(SNPassoc)
library(aod)
```

## Dataset preparation

ДЛЯ ДАТЫ МОЗГОВОЙ КАТАСТРОФЫ для пациента 2 пацентов указан только год
для пациента (170 строка) указан год и дата через дробь??? 9 пропущенных
значений NA

ДЛЯ ДАТЫ ПЕРЕВОДА В ФНКРР для пациента 21 опечатка 6 NA

Read dataset and check for errors in date variables

```{r cars}
data <- read_csv2("data.csv")

data$`Дата перевода в ФНКРР`[191] <- "09.11.2018"
data$`Дата мозговой катастрофы`[8] <- "01.01.2014"
data$`Дата мозговой катастрофы`[169] <- "12.09.2017"
data$`Дата мозговой катастрофы`[321] <- "01.01.2014"

data <- data %>% mutate(`Дата мозговой катастрофы` = dmy(`Дата мозговой катастрофы`),
                     `Дата перевода в ФНКРР` = dmy(substr(`Дата перевода в ФНКРР`, start=1, stop=8)),
                     `Дата выписки из ФНКЦ РР` = dmy(`Дата выписки из ФНКЦ РР`)
                     )

```
```{r}
######## Replace нд and normalize variable `внк ДНК`
data <- data %>%
  mutate(across(everything(), ~na_if(as.character(.), "н/д")))

#data <- data %>%
  #mutate(`Дата мозговой катастрофы` = dmy(`Дата мозговой катастрофы`))
target_columns <- c("внкДНК, нг/мл 1d", "8ohdg на миллион оснований 1d",
                    "внкДНК, нг/мл 5d", "8ohdg на миллион оснований 5d")


data <- data %>%
  mutate(across(all_of(target_columns), as.numeric))


data <- data %>%
  mutate(
    oneday_log = if_else(`внкДНК, нг/мл 1d` > 0, log(`внкДНК, нг/мл 1d`), NA_real_),
    fiveday_log = if_else(`внкДНК, нг/мл 5d` > 0, log(`внкДНК, нг/мл 5d`), NA_real_)
  )


data1 <- data %>% slice(1:173)
data2 <- data %>% slice(220:522)

mean1 <- mean(data1$oneday_log, na.rm = TRUE)
mean2 <- mean(data2$oneday_log, na.rm = TRUE)

coeff1 <- mean2 / mean1

data <- data %>%
  mutate(oneday_log = if_else(row_number() <= 173, 
                              oneday_log * coeff1, 
                              oneday_log))

data1 <- data %>% slice(1:173)
data2 <- data %>% slice(220:522)

mean1 <- mean(data1$fiveday_log, na.rm = TRUE)
mean2 <- mean(data2$fiveday_log, na.rm = TRUE)

coeff2 <- mean2 / mean1


data <- data %>%
  mutate(fiveday_log = if_else(row_number() <= 173, 
                               fiveday_log * coeff2, 
                               fiveday_log))


data

###### 8ohg 
data <- data %>%
  mutate(
    `8ohdg 1d log` = if_else(`8ohdg на миллион оснований 1d` > 0, log(`8ohdg на миллион оснований 1d`), NA_real_),
    `8ohdg 5d log` = if_else(`8ohdg на миллион оснований 5d` > 0, log(`8ohdg на миллион оснований 5d`), NA_real_)
  )

data1 <- data %>% slice(1:173)
data2 <- data %>% slice(220:522)

mean1_ohdg1 <- mean(data1$`8ohdg 1d log`, na.rm = TRUE)
mean2_ohdg1 <- mean(data2$`8ohdg 1d log`, na.rm = TRUE)

coeff_ohdg1 <- mean2_ohdg1 / mean1_ohdg1

data <- data %>%
  mutate(`8ohdg на миллион оснований 1d лог` = if_else(row_number() <= 173, 
                                                      `8ohdg 1d log` * coeff_ohdg1, 
                                                      `8ohdg 1d log`))

mean1_ohdg5 <- mean(data1$`8ohdg 5d log`, na.rm = TRUE)
mean2_ohdg5 <- mean(data2$`8ohdg 5d log`, na.rm = TRUE)

coeff_ohdg5 <- mean2_ohdg5 / mean1_ohdg5

data <- data %>%
  mutate(`8ohdg 5d лог` = if_else(row_number() <= 173, 
                                                      `8ohdg 5d log` * coeff_ohdg5, 
                                                      `8ohdg 5d log`))

data
data_e <- data
```
 List of primary variables (according to medical basis) : 
  Номер, ID,  Пол, HTR1A_rs6295, AQP4_rs1058427 , BDNF_rs6265, OLR1_rs11053646, NRF2_rs6726395, AGTR1_rs275651, AQP5_rs3759129, TLR9_rs352162, TLR9_rs187084, AQP_rs3736309, Возраст, анти-HCV, Hbs_AG,  ЛСМА, Основной_диагноз_код_МКБ10, Диагноз, Пневмония, Инфекция, CGS_балл, Индекс_коморбидности_Чарлсона_1d, CIRS-G_сумма_1d, CIRS-G_систем_с_паталогиями_1, CIRS-G_индекс_тяжести_1d, FOUR_1d, SOFA_1d, внкДНК_1d, 8ohdg_1d, WBC_1d,  NEU_#_1d, NEU_%_1d, LYM_#_1d, LYM_%_1d,  PLT_1d, Лактат_1d, Креатинин_1d, АЛТ_1d, АСТ_1d,  Альбумин_1d, Щелочная_фосфотаза_1d , Мочевая_кислота_1d, Мочевина_1d, ГГТ_1d, Калий_1d, Натрий_1d, Хлорид_1d,  АЧТВ_1d, МНО_1d, Фибриноген_1d, ПВ_1d, Уросептическая_инфекция, Сепсис/септический_шок, Потребность_ИВЛ, Ухудшение_пневмонии_сутки, ИВЛ_д, Дни_ОРИТ_ФНКЦРР, Койко-дни_ФНКЦРР, Исход, Пневмония_возникшая_ФНКЦРР 

#### ENGLISH RENAME

```{r}
data_e <- data %>% rename ("Num" = "№ истории", "Part" = "Часть", "Sex" = "Пол", "Num_ref" = "Номер Настя", "Age" = "Возраст", "HTR1A_rs6295" = "HTR1A  rs6295","AQP4_rs1058427"  = "AQP4 rs1058427", "BDNF_rs6265" = "BDNF rs6265", "OLR1_rs11053646" = "OLR1 rs11053646", "NRF2_rs6726395" = "NRF2 rs6726395", "AGTR1_rs275651" = "AGTR1  rs275651",  "AQP5_rs3759129" = "AQP5 rs3759129", "TLR9_rs352162" = "TLR9 rs352162", "TLR9_rs187084" = "TLR 9 rs187084", "AQP5_rs3736309" = "AQP rs3736309",  "anti-HCV" = "анти HCV 2-нд, 1 - да, 0 - нет", "Hbs_AG" = "Hbs AG 0-не,2-нд, 1 - да", "LMCA" = "ЛСМА", "Date_brain_damage" = "Дата мозговой катастрофы" , "Main_diagnosis_code" = "Основной диагноз код МКБ10", "Diagnosis" = "Диагноз целиком", "Transfer_date_FSCCRR" = "Дата перевода в ФНКРР", "Pneumonia" = "Поступил в ФНКЦ с пневмонией", "Infection"  = "Поступил в ФКНЦ с инфекцией", "T_admission"  = "t° при поступлении", "CGS" = "CGS, балл", "CCI_1d"  = "Индекс коморбидности Чарлсона (баллы) 1d",  "CCI_5d" = "Индекс коморбидности Чарлсона (баллы) 5d",
"CCI_21d" = "Индекс коморбидности Чарлсона (баллы) 21d",
"CCI_out" = "Индекс коморбидности Чарлсона (баллы)  out",
"CIRS-G_sum_1d"  = "CIRS-G сумма, 1d",
"CIRS-G_pathology_1d" = "CIRS-G систем с паталогиями, 1d" ,
"CIRS-G_severity_1d" = "CIRS-G индекс тяжести, 1d",
"CIRS-G_sum_5d_35"  = "CIRS-G сумма, 5d...35",
"CIRS-G_pathology_5d_36"= "CIRS-G систем с паталогиями, 5d...36",
"CIRS-G_severity_5d_37" = "CIRS-G индекс тяжести, 5d...37",
"CIRS-G_sum_5d_38" = "CIRS-G сумма, 5d...38",
"CIRS-G_pathology__5d_39"  = "CIRS-G систем с паталогиями, 5d...39",
"CIRS-G_severity_5d_40" = "CIRS-G индекс тяжести, 5d...40",
"CIRS-G_sum_out" = "CIRS-G сумма, out",
"CIRS-G_pathology_out" = "CIRS-G систем с паталогиями, out",
"CIRS-G_severity_out" = "CIRS-G индекс тяжести, out",
"exDNA_1d" = "внкДНК, нг/мл 1d",
"8ohdg_1d" = "8ohdg на миллион оснований 1d",
"exDNA_5d" = "внкДНК, нг/мл 5d",
"8ohdg_5d" = "8ohdg на миллион оснований 5d",
"Uroseptic_infection" = "Уросептическая инфекция в любой момент времени",
"Sepsis/septic_shock" = "Сепсис/септический шок в любой момент времни",
"Ventilation" = "Потребность в ИВЛ за время госпитализации в любой момент времени",
"Ухудшение пневмонии_сутки" = "Ухудшение исходной пневмонии, сутки",
"Ventilation_d" = "ИВЛ (дни)",
"Discharge_date_FSCCRR" = "Дата выписки из ФНКЦ РР",
"ICU_FSCCRR_d" = "Дней в ОРИТ ФНКЦ РР",
"FSCCRR_bed-days" = "Койко-дней в ФНКЦ РР",
"Outcome" = "Исход, 0 -выписан, 1 умер, 2 - цензура",
"Pneumonia_at_FSCCRR" = "Пневмония возникшая в ФНКЦ РР",
 "Pneuminia_deterioration_d" = "Ухудшение исходной пневмонии, сутки") %>% 
  rename_with(~ gsub("Лактат","Lactate", .x)) %>% 
  rename_with(~ gsub("Креатинин","Creatinine", .x)) %>% 
  rename_with(~ gsub("АЛТ","ALT", .x)) %>% 
  rename_with(~ gsub("АСТ","AST", .x)) %>% 
  rename_with(~ gsub("Альбумин","Albumin", .x)) %>% 
  rename_with(~ gsub("Щелочная фосфотаза","Alcaline_phosphatase", .x)) %>% 
  rename_with(~ gsub("Мочевая кислота","Uric_acid", .x)) %>% 
  rename_with(~ gsub("Мочевина","Urea", .x)) %>% 
  rename_with(~ gsub("ГГТ","GGT", .x)) %>% 
  rename_with(~ gsub("Натрий","Sodium", .x)) %>% 
  rename_with(~ gsub("Калий","Potassium", .x)) %>% 
  rename_with(~ gsub("Хлорид","Chloride", .x)) %>% 
  rename_with(~ gsub("АЧТВ","APTT", .x)) %>% 
  rename_with(~ gsub("МНО","INR", .x)) %>% 
  rename_with(~ gsub("Фибриноген","Fibrinogen", .x)) %>% 
  rename_with(~ gsub("ПВ","PT", .x)) %>% 
  rename_at(vars(44 : 148), ~str_replace_all(., "\\s+", "_"))  %>% 
  mutate(across(`Age`, as.integer)) %>% 
  # delete initial dna- variables  
  select (!c(`8ohdg на миллион оснований 1d лог`, `8ohdg 5d лог`))

colnames(data_e)
```

Calculate term before transfer to FSCCRR (`Before_FSCCRR_d`) and clean outcome variables (н/д replace for NA, Outcome 2 - for NA), also check and correct letter case for sex designation

Generate new variable NLR_1d


```{r }

data_e <- data_e %>% mutate (`Before_FSCCRR_d` = as.Date(data_e$`Transfer_date_FSCCRR` , format="%d-%m-%Y")- as.Date(data_e$`Date_brain_damage`, format="%d-%m-%Y"))%>% 
  mutate(`Ventilation` = ifelse(`Ventilation` =="н/д", NA, `Ventilation`), `Outcome` = ifelse( `Outcome` == 2, NA, `Outcome`),
 `Sex` = case_when(`Sex` %in% c('ж', 'Ж') ~ 'F', `Sex`  %in% c('м', 'М') ~ 'M', .default = `Sex`),
 `Pneumonia` = case_when(`Pneumonia` == 1 ~ "Yes", `Pneumonia` == 0 ~ "No", TRUE ~ NA),
 `Infection` = case_when(grepl("1",`Infection` ) ~ "Yes", `Infection` == 0 ~ "No", TRUE ~ NA),
 `Sepsis/septic_shock` = case_when(grepl("1",`Sepsis/septic_shock` ) ~ "Yes", `Sepsis/septic_shock` == 0 ~ "No", TRUE ~ NA)) %>% 
  mutate (`NLR_1d` = as.numeric(`NEU_#_1d`) / as.numeric(`LYM_#_1d`))

```

Note about NLR:
NLR reflects online dynamic relationship between innate (neutrophils) and adaptive cellular immune response (lymphocytes) during illness and various pathological states. NLR is influenced by many conditions including age, rice, medication, chronic disease like coronary heart disease, stroke, diabetes, obesity, psychiatric diagnosis, cancer of solid organs, anemia and stress. A normal range of NLR is between 1-2, the values higher than 3.0 and below 0.7 in adults are pathological. NLR in a grey zone between 2.3-3.0 may serve as early warning of pathological state or process such like cancer, atherosclerosis, infection, inflammation, psychiatric disorders and stress. 



### SNP preparation

SNP Preparation : "HTR1A_rs6295" "AQP4_rs1058427" "BDNF_rs6265"
"OLR1_rs11053646" "NRF2_rs6726395" "AGTR1_rs275651" "AQP5_rs3759129"
"TLR9_rs352162" "TLR9_rs187084" "AQP5_rs3736309"


Generate separate dataset with corrected SNP, outcomes (`Ventilation`, `Outcome`), DNA variables ( `oneday_log`, `8ohdg 1d log`) and NLR variable.

```{r}

hospital_polymorf_e <- data_e %>%  mutate (`HTR1A_rs6295` = ifelse (`HTR1A_rs6295` == 'С/С', 'C/C', `HTR1A_rs6295`),
                              `AQP4_rs1058427` = ifelse (`AQP4_rs1058427` == 'C/A'| `AQP4_rs1058427` == 'С/A', 'C/A', `AQP4_rs1058427`),
                              `OLR1_rs11053646` = ifelse (`OLR1_rs11053646` == 'C/G', 'G/C', `OLR1_rs11053646`),
                              `AQP5_rs3759129` = ifelse (`AQP5_rs3759129` == 'С/C', 'C/C',`AQP5_rs3759129`),
                              `TLR9_rs352162` = case_when(`TLR9_rs352162` == 'С/С' | `TLR9_rs352162` == 'С/C' ~ 'C/C',
                                                           
                                                        `TLR9_rs352162` == 'С/T' ~ 'C/T',
                                                        .default = `TLR9_rs352162` ),
                              `TLR9_rs187084` = case_when(`TLR9_rs187084` == 'C/T' | `TLR9_rs187084` == 'С/T' ~ 'C/T',
                                                        .default = `TLR9_rs187084` ),
                              `AQP5_rs3736309` = case_when(`AQP5_rs3736309` == 'C/T' | `AQP5_rs3736309` == 'С/T' ~ 'C/T',
                                                        .default = `AQP5_rs3736309` ))  %>% 
  select(ID, HTR1A_rs6295, AQP4_rs1058427, BDNF_rs6265, OLR1_rs11053646, NRF2_rs6726395, AGTR1_rs275651, AQP5_rs3759129, TLR9_rs352162, TLR9_rs187084, AQP5_rs3736309, `Ventilation`, `Outcome`, `oneday_log`, `8ohdg 1d log`, `NLR_1d` ) %>% 
  mutate (across(c(HTR1A_rs6295, AQP4_rs1058427, BDNF_rs6265, OLR1_rs11053646, NRF2_rs6726395, AGTR1_rs275651, AQP5_rs3759129, TLR9_rs352162, TLR9_rs187084, AQP5_rs3736309), factor)) 

```

SNP distribution graph

```{r}
hospital_polymorf_long <- hospital_polymorf_e %>%
  select (ID, HTR1A_rs6295, AQP4_rs1058427, BDNF_rs6265, OLR1_rs11053646, NRF2_rs6726395, AGTR1_rs275651, AQP5_rs3759129, TLR9_rs352162, TLR9_rs187084, AQP5_rs3736309) %>%
  pivot_longer(!ID, names_to = "Position", values_to = "SNP")

ggplot (hospital_polymorf_long) +
  geom_bar (aes(SNP)) +
  facet_wrap (~Position, nrow = 4, axes = "all", axis.labels = "all_x")
```


**SNP descriptive analysis**

```{r}
library(SNPassoc)
# specify columns contain SNP
polymorf.s <- setupSNP (data=hospital_polymorf_e, colSNPs=2:11, sep="/")
# got new class snp

head(polymorf.s$TLR9_rs352162)
summary(polymorf.s$TLR9_rs352162)
# which shows the genotype and allele frequencies for a given SNP, testing for Hardy-Weinberg equilibrium (HWE).
```

```{r}
 #We can  visualize the results in a plot by
plot(polymorf.s$TLR9_rs352162)
plot(polymorf.s$TLR9_rs352162, type=pie)
```


Genotyping of SNPs needs to pass quality control measures. Aside from technical details that need to be considered for filtering SNPs with low quality, genotype calling error can be detected by a HWE test. The test compares the observed genotype frequencies with those expected under random mating, which follows when the SNPs are in the absence of selection, mutation, genetic drift, or other forces. Therefore, HWE must be checked only in controls. There are several tests described in the literature to verify HWE. In SNPassoc HWE is tested for all the bi-allelic SNP markers using a fast exact test Wigginton, Cutler, and Abecasis (2005) implemented in the tableHWE function.

```{r}
# results across all snp
summary(polymorf.s, print=FALSE)
```


Showing the SNP labels with minor/major allele format, the major allele
frequency the HWE test and the percentage of missing genotypes. Missing
values can be further explored plotting with:

```{r}
plotMissing(polymorf.s, print.labels.SNPs = FALSE)
```

Missing genotypes. Black squares shows missing genotype information of dataset.
This plot can be used to inspect if missing values appear randomly across individuals and SNPs. In our case, we can see that the missing pattern may be considered not random. Group of individuals should be further checked for possible problems with genotyping.
MNCR?


**Hardy-Weinberg equilibrium**

Genotyping of SNPs needs to pass quality control measures. Aside from technical details that need to be considered for filtering SNPs with low quality. The test compares the observed genotype frequencies with those expected under random mating, which follows when the SNPs are in the absence of selection, mutation, genetic drift, or other forces. Therefore, HWE must be checked only in controls. There are several tests described in the literature to verify HWE. In SNPassoc HWE is tested for all the bi-allelic SNP markers using a fast exact test Wigginton, Cutler, and Abecasis (2005) implemented in the tableHWE function.

Set Ventillation demand as a Outcome for further HWE analysis.

```{r}
hwe <- tableHWE(polymorf.s)
hwe
```
We observe that TLR9_rs352162 SNPs in the dataset is not under HWE since their P-values rejecting the HWE hypothesis (null hypothesis) are less than 0.05. However, when tested in control samples only (appreciated attitude), by stratifying by cases and controls:

Check  HWE in control `Outcome` ("0" - discharge)
```{r}
hwe2 <- tableHWE(polymorf.s, `Outcome`)

#SNPs is HWE in the whole sample but not controls!!
snpNHWE <- hwe2[,1]>0.05 & hwe2[,2]<0.05
rownames(hwe2)[snpNHWE]
```

In case  TLR9_rs187084, reject null hypotesis HWE (p-value for control group  < 0.05). Otherwise one is interested in keeping those SNPs that **do not reject the null hypothesis**. 

NOTE: As several SNPs are tested, multiple comparisons must be considered. In this particular setting, a threshold of 0.001 is normally considered. As a quality control measure, it is not necessary to be as conservative as in those situations where false discovery rates need to be controlled.

HWE analysys considered TLR9_rs352162 and TLR9_rs187084 are not under HWE and should be excluded, but alternative casecontrol stratification (by Ventilation) showed that all SNP are under HWE in control group:

```{r}
hwe3 <- tableHWE(polymorf.s, `Ventilation`)

#SNPs is HWE in the whole sample but not controls
snpNHWE <- hwe3[,1]>0.05 & hwe3[,2]<0.05
rownames(hwe3)[snpNHWE]
```

The decision was to assign all SNP to furter analysis.

```{r}
#SNPs that do not pass the HWE test must be removed form further analyses. We can recall setupSNP and indicate the columns of the SNPs to be kept

#snps.ok <- rownames(hwe2)[hwe2[,2]>=0.001]
#pos <- which(colnames(hospital_polymorf_e)%in%snps.ok, useNames = FALSE)
#polymorph.s <- setupSNP(hospital_polymorf_e, pos, sep="/")
```


**SNP association analysis**
**Outcome analysis**

We are interested in finding those SNPs associated with outcome status. The association analysis for all genetic models is performed by the function association that regresses Outcome on the variable SNP from the dataset that already contains the SNP variables of class snp.

```{r}
association(`Outcome` ~ TLR9_rs352162, data= polymorf.s)
```
The P-values obtained from massive univariate analyses are visualized with the generic plot function

```{r}
ans <- WGassociation(`Outcome`, data=polymorf.s)
ans
plot(ans)
```
Manhattan-type plots for different genetic models. P-values in -log10 scale to assess the association between case-control status and SNPs in the asthma example.
This produces a Manhattan plot of the -log10(P-values) for all the SNPs over all models. It shows the nominal level of significance and the Bonferroni level, which is the level corrected by the multiple testing across all SNPs. The overall hypothesis of massive univariate association tests is whether there is any SNP that is significantly associated with the phenotype. As multiple SNPs are tested, the probability of finding at least one significant finding increases if we do not lower the significance threshold. The Bonferroni correction lowers the threshold by the number of SNPs tested (0.0001=0.05/51). In the Manhattan plot of our analysis, we see that no SNP is significant at the Bonferroni level, and therefore there is no SNP that is significantly associated with Outcome.

```{r}
association(`NLR_1d` ~ `AQP4_rs1058427` , data= polymorf.s)
```


```{r}
ans_nlr <- WGassociation(`NLR_1d`, data=polymorf.s)
ans_nlr
plot(ans_nlr)
```

```{r}
ans_dna <- WGassociation(`oneday_log`, data=polymorf.s)
ans_dna
plot(ans_dna)
```


### Create new variable "deciase group" for classification of diagnoses codes


```{r}
data_e <- data_e %>% mutate(
    `Main_diagnosis_code` = case_when(
    `Main_diagnosis_code` == "I 69.1" ~ "I69.1",
    `Main_diagnosis_code` == "I 69.3" ~ "I69.3",
    `Main_diagnosis_code` == "G.93.1" ~ "G93.1",
    `Main_diagnosis_code` == "I.69." ~ "I69",
    `Main_diagnosis_code` == "I.69.3" ~ "I69.3",
    `Main_diagnosis_code` == "Т.94." ~ "Т94",
    `Main_diagnosis_code` == "I 69.0" ~ "I69.0",
    `Main_diagnosis_code` == "Т90." ~ "Т90",
    `Main_diagnosis_code` == "I.21.4" ~ "I21.4",
    `Main_diagnosis_code` == "I.69.1" ~ "I69.1",
    `Main_diagnosis_code` == "Т 90.5" ~ "Т90.5",
    `Main_diagnosis_code` == "D33.0 D" ~ "D33.0",
    `Main_diagnosis_code` == "M86.1 О" ~ "M86.1",
    `Main_diagnosis_code` == "G93.1 G" ~ "G93.1",
    `Main_diagnosis_code` == "67.9 Це" ~ "I67.9",
    `Main_diagnosis_code` == "G93.1 А" ~ "G93.1",
    `Main_diagnosis_code` == "G09 ос" ~ "G09",
    `Main_diagnosis_code` == "I69.0 1" ~ "I69.0",
    `Main_diagnosis_code` == "G. 93.1" ~ "G93.1",
    `Main_diagnosis_code` == "26" ~ "I69.3",              #!!!!
    `Main_diagnosis_code` == "G93..4" ~ "G93.4",
    `Main_diagnosis_code` == "169.3"	~ "I69.3",
    `Main_diagnosis_code` == "169.1"	~ "I69.1",
    `Main_diagnosis_code` == "160.0"	~ "I60.0",
    `Main_diagnosis_code` == "169.4" ~ "I69.4",
    `Main_diagnosis_code` == "T90,5" ~ "T90.5",
    `Main_diagnosis_code` == "T.94.0" ~ "T94.0",
    TRUE ~ `Main_diagnosis_code`
  ),
  `Main_diagnosis_code` = str_replace_all(`Main_diagnosis_code`, "С", "C"),
  `Main_diagnosis_code` = str_replace_all(`Main_diagnosis_code`, "Т", "T"),
  `Main_diagnosis_code` = str_replace_all(`Main_diagnosis_code`, ";", ""),
  `Main_diagnosis_code` = str_replace_all(`Main_diagnosis_code`, ", ", ";"),
  `Main_diagnosis_code` = str_replace_all(`Main_diagnosis_code`, "    ", " "),
  `Main_diagnosis_code` = str_replace_all(`Main_diagnosis_code`, " ", ";"),
)
```

```{r}
data_e %>% summarise(
  diseaseGroup = case_when(
    `Main_diagnosis_code` %in% c("I63", "I69.3", "I69.4", "T82.8") ~ "Ишемический инсульт",
    `Main_diagnosis_code` %in% c("Q28", "I60", "I72", "I69.2", "I69.1", "I69.0", "I61") ~ "Геморрагический инсульт",
    `Main_diagnosis_code` %in% c("S06", "T94", "T94") ~ "Последствие ЧМТ",
    `Main_diagnosis_code` %in% c("G93.4", "G93.1", "G94.3", "I46", "U07.1", "U09.9", "G96", "I21", "I23.1") ~ "Гипоксия",
    `Main_diagnosis_code` %in% c("С71", "D33", "D32", "D43", "I72") ~ "Новообразование",
    TRUE ~ "OTHER"
  ), `Main_diagnosis_code`) %>% filter(diseaseGroup == "OTHER") %>% unique()
```


```{r}
# Классифицируем диагнозы по группам
data_e <- data_e %>% mutate(
  diseaseGroup = case_when(
    `Main_diagnosis_code` %in% c("I63", "I69.3", "I69.4", "T82.8", "I69.3;C34.1", "I63.5", "I63.3", "I63.2", "I63.4", "I63.9") ~ "Ischemic stroke",
    `Main_diagnosis_code` %in% c("Q28", "I60", "I72", "I69.2", "I69.1", "I69.0", "I61", "I60.2", "I61.6", "I61.5", "I61.3", "I60.3", "I60.6", "I60.0") ~ "Hemorrhagic stroke",
    `Main_diagnosis_code` %in% c("S06", "T94", "T90", "T90.4", "T94.0", "T94", "T90.1", "T90", "T90.2", "T90.5", "T90.8", "S06.6", "S06.31", "S06.2;T88.8", "T94.1") ~ "TBI",
    `Main_diagnosis_code` %in% c("G93.4", "G93.1", "G94.3", "I46", "U07.1", "U09.9", "G96", "I21", "I23.1", "I46;G93.1", "I21.4", "G94.3;G82.4") ~ "Hypoxia",
    `Main_diagnosis_code` %in% c("С71", "D33", "D32", "D43", "I72", "C71.2", "D33.3", "D33.1", "D33.0", "D43.0;G81.9", "D43.1", "D43.0", "D32.2", "D32.0", "C71.1", "I72.0", "I72.8") ~ "Tumor",
    `Main_diagnosis_code` %in% c("I67.2", "I67.1") ~ "Ischemic stroke",
    `Main_diagnosis_code` %in% c("I69.0", "I69.0;G91.0") ~ "Hemorrhagic stroke",
    `Main_diagnosis_code` %in% c("T06.8", "T91.3", "T98.1", "T82.8") ~ "TBI",
    `Main_diagnosis_code` %in% c("G93.1", "G82.4", "S15.0", "J17.1") ~ "Hypoxia",
    `Main_diagnosis_code` %in% c("D35.3;G04.2", "D35.2", "C79.3", "C71.8", "C71.6", "C71") ~ "Tumor",
    TRUE ~ NA
  ),
  `Main_diagnosis_code`
) 


data_e <- data_e %>% mutate(diseaseGroup = as.factor(diseaseGroup))
# Find ungroped diagnosis
data_e %>%
  select(`Main_diagnosis_code`, diseaseGroup) %>%
  filter(is.na(diseaseGroup))  %>%  unique() %>% arrange(.desc=T)

```

```{r fig.height=15, fig.width=17}
data_e %>%
  group_by(diseaseGroup) %>%
  drop_na(diseaseGroup) %>% 
  summarise(count = n()) %>% 
  ggplot(aes(x = diseaseGroup, y = count)) +
  geom_bar(stat="identity", col = 'black', fill = 'white', linewidth=1) +
  geom_label(aes(label=count), col = 'black', size=10, label.size=1) +
  scale_x_discrete('Disease group') +
  scale_y_continuous('N patients') +
  theme(
    axis.text.x = element_text(size = 21, face = 'bold'),
    axis.text.y = element_text(size = 23),
    axis.title = element_text(size = 35),
    legend.title = element_text(size = 20),
    legend.text = element_text(size = 20),
    panel.background = element_rect(fill = "white", color = 'black'),
    plot.background = element_rect(fill = "white"),
    panel.grid.major = element_line(color = "gray"),
    panel.grid.minor = element_line(color = "lightgray"),
  )
```
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!!!!!!!!НАЧИНАЯ ОТСЮДА Я НЕ ИСПРАВЛЯЛА ЧАСТЬ УСТИНА!!!!!!!!!!!!!!!!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!


```{r}
data_no_na_dates <- data %>% filter(!is.na(`Дата_выписки_ФНКЦРР`)) %>% filter(!is.na(`Дата_выписки_ФНКЦРР`))

forcrmp <- data_no_na_dates %>%
  mutate(
    time = `Койко-дни_ФНКЦРР` %>% as.numeric(),
    status = case_when(
      `Исход` == "0" ~ "Выписка",
      `Исход` == "1" ~ "Смерть",
      `Исход` == "2" ~ "Цензура"
    ) %>% factor(levels = c("Цензура", "Смерть", "Выписка"))
  ) %>% 
  select(time, status, `Уросептическая_инфекция`, diseaseGroup, Исход)
```


## Competing risk regression
```{r}
crr_mod <- crr(Surv(time, status) ~ `Уросептическая_инфекция`, forcrmp)
crr_mod
```

```{r}
tbl <- 
  crr_mod %>%
  gtsummary::tbl_regression(exponentiate = TRUE) %>%
  gtsummary::add_global_p() %>%
  add_n(location = "level")
```
## Cumulative incidence
```{r}
cuminc(Surv(time, status) ~ 1, forcrmp)
```

```{r}
cuminc(Surv(time, status) ~ 1, forcrmp) %>%
  ggcuminc(c('Смерть', 'Выписка')) +
  add_confidence_interval('ribbon') +
  add_risktable(risktable_stats = c("n.risk", "cum.event", 'cum.censor')) +
  scale_ggsurvfit(x_scales = list(breaks = seq(0, 600, by = 30)))
```


```{r}
cuminc(Surv(time, status) ~ diseaseGroup, forcrmp)
```

```{r}
cuminc(Surv(time, status) ~ diseaseGroup, forcrmp) %>%
  ggcuminc(c('Смерть', 'Выписка')) +
  scale_ggsurvfit(x_scales = list(breaks = seq(0, 600, by = 30)))+
  add_risktable()
```

### L Patients status at FSCCRR delivery and demographics

```{r fig.width= 10, fig.height=8}

theme_cust <- theme(
    axis.text.x = element_text(size = 14, face = 'bold'),
    axis.text.y = element_text(size = 14),
    axis.title = element_text(size = 14),
    legend.title = element_text(size = 14),
    legend.text = element_text(size = 14),
    panel.background = element_rect(fill = "white", color = 'black'),
    plot.background = element_rect(fill = "white"),
    panel.grid.major = element_line(color = "gray90"),
    panel.grid.minor = element_line(color = "lightgray"),
    strip.text = element_text(size = 14)
  )

#SEX
ggplot(data_e %>% filter(!is.na(Sex))) +
  geom_bar(aes(y = diseaseGroup))+
  facet_wrap(~`Sex`, ncol = 1) +
  scale_y_discrete('Disease group') +
  scale_x_continuous('Number of patients') +
  theme_light() +
  theme_cust
  

#AGE

ggplot(data_e) +
  geom_histogram (aes(`Age`)) + 
  facet_wrap(~diseaseGroup) +
  scale_y_discrete('Disease group') +
  scale_x_continuous('Number of patients') +
  theme_light() +
  theme_cust

ggplot(data_e %>% filter(!is.na(Sex))) +
  geom_pointrange(mapping = aes(y = diseaseGroup , x = Age),
                  stat = "summary",
                  fun.min = function(z) {quantile(z,0.25)},
                  fun.max = function(z) {quantile(z,0.75)},
                  fun = median, 
                  size = 1) +
  facet_wrap(~`Sex`, ncol = 1) +
  scale_y_discrete('Disease group') +
  scale_x_continuous('Number of patients') +
  theme_light() +
  theme_cust


#PNEUMONIA 

ggplot(data_e) +
  geom_bar (aes(Pneumonia)) +
  facet_wrap(~diseaseGroup) +
  scale_y_continuous('Number of patients') +
  theme_light() +
  theme_cust


#INFECTION AT TRANSFER
ggplot(data_e) +
  geom_bar (aes(`Infection`)) +
  facet_wrap(~diseaseGroup) +
  scale_y_continuous('Number of patients') +
  theme_light() +
  theme_cust

#SOFA
ggplot(data_e) +
  geom_bar (aes(`SOFA_1d`)) +
  facet_wrap(~diseaseGroup) +
  scale_x_discrete('SOFA 1 day score') +
  scale_y_continuous('Number of patients') +
  theme_light() +
  theme_cust

#CCI
ggplot(data_e) +
  geom_bar (aes(`CCI_1d`)) +
  facet_wrap(~diseaseGroup) +
  scale_x_discrete('CCI 1 day') +
  scale_y_continuous('Number of patients') +
  theme_light() +
  theme_cust
```


### Univariate regression for all variables (at time point 0)

Prepate ataset for regression preparation, including 0-day status, demographic data, disease group and NLR. Also create a new categorial variable for bed-days at FSCCRR based on quantiles , called **BD_group**

```{r}

# calculate quantiles for bed-days  categorization 

BD <- as.numeric(data_e$`FSCCRR_bed-days`)
res<-quantile(BD, probs = c(0,0.25,0.5,0.75,1), na.rm = T)
res

data_for_regres <- data_e %>% select (ID, Sex, Age, `anti-HCV`, `Hbs_AG`,  `LMCA`, Pneumonia, Infection, diseaseGroup, CGS, CCI_1d,`CIRS-G_sum_1d`, `CIRS-G_pathology_1d`, `CIRS-G_severity_1d`, FOUR_1d, SOFA_1d, exDNA_1d, `8ohdg_1d`, WBC_1d,  `NEU_#_1d`, `LYM_#_1d`,`PLT_1d`, `Lactate_1d`, Creatinine_1d, ALT_1d, AST_1d,  Albumin_1d, Alcaline_phosphatase_1d , Uric_acid_1d, Urea_1d, GGT_1d, Potassium_1d, Sodium_1d, Chloride_1d,  APTT_1d, INR_1d, Fibrinogen_1d, PT_1d, `Uroseptic_infection`, `Sepsis/septic_shock`, `Ventilation`, `Pneuminia_deterioration_d`, `Ventilation_d`, `ICU_FSCCRR_d`, `FSCCRR_bed-days`, `Outcome`, `Pneumonia_at_FSCCRR`, `oneday_log`, `8ohdg 1d log`) %>% 
  
  mutate(across(c(CGS, CCI_1d ,`CIRS-G_sum_1d`, `CIRS-G_pathology_1d`,`CIRS-G_severity_1d`,`FOUR_1d`, `SOFA_1d`, `WBC_1d`, `NEU_#_1d`, `LYM_#_1d`, `PLT_1d`, `Lactate_1d`, `Creatinine_1d` , `ALT_1d`, `AST_1d`, `Albumin_1d`, `Alcaline_phosphatase_1d`, `Uric_acid_1d`, `Urea_1d` , `GGT_1d`, `Potassium_1d`, `Sodium_1d`, `Chloride_1d`, `APTT_1d`, `INR_1d`,`Fibrinogen_1d`, PT_1d , `Pneuminia_deterioration_d`, `Ventilation_d`, `ICU_FSCCRR_d`, `FSCCRR_bed-days`), as.numeric), 
         across(c(Outcome,`Ventilation`), factor)) %>%
  # new variable - categorised `FSCCRR_bed-days` by quantiles
  mutate (BD_group = case_when (`FSCCRR_bed-days` < 30 ~ "0-30",
                                30 <= `FSCCRR_bed-days` & `FSCCRR_bed-days` < 46 ~ "30-46",
                                46 <= `FSCCRR_bed-days` & `FSCCRR_bed-days` < 67 ~ "46-67",
                                67 < `FSCCRR_bed-days` ~ "67-401"))
  

# add corrected SNP data

hosp_poly_group_nlr  <- left_join(data_for_regres, hospital_polymorf_e[,1:11], by = "ID" )  %>%  select (!ID) %>% filter(!is.na(Ventilation))

#Achieve corrected  dataset with right names and correct variables types

#summary(hosp_poly_group_nlr)
#str(hosp_poly_group_nlr)
#colnames(hosp_poly_group_nlr)

hosp_poly_group_nlr %>% group_by(`BD_group`)  %>% tally ()

```

**Based on clinical data and analysis of cases count it was decided to chose only discharged patients (exclude '1'/dead cases) and evaluate ventilation demand as an outcome for analysis **

```{r}

discharged <- hosp_poly_group_nlr %>% filter(Outcome == "0") %>%  select (!c(Outcome, `Ventilation_d`))
discharged$Ventilation
```

 Run regression for all variables of interest

```{r}
x <- names(discharged[,-40])
formulas <- unlist(lapply(40, function(n) combn(x, 1, FUN=function(row) paste0("Ventilation ~ ", paste0("`",row, "`",collapse = "+")))))
formulas  # combine each variable into formula ("outcome ~ variable of interest")
```

```{r warning=FALSE}
library(broom)
library(openxlsx)


#regression coefficients
tmp1 <- bind_rows(lapply(formulas, function(frml) {
 a <- tidy(glm(frml, data= discharged, family = binomial), exponentiate = TRUE, conf.int = TRUE) %>%
   mutate(across(where(is.numeric), round, digits = 2))
 a$frml = frml
 return(a)
}))

tmp1 <- tmp1 %>% mutate (`sign_.05` = ifelse (`p.value` >= 0.05, "no", "yes" ))

signif_estim_univar <- as_data_frame(tmp1) %>% filter (`p.value` <= 0.05)

write.xlsx(signif_estim_univar, 'signif_estim_univar_glm.xlsx')
#regression results R2, AIC, BIC
#tmp2 = bind_rows(lapply(out, function(frml) {
# a = glance(lm(frml, data=hosp_poly_group_nlr))
# a$frml = frml
# return(a)
# }))
# tmp2
#signif_results_univar<- as_data_frame(tmp2)
#write.xlsx(signif_estim_univar, 'signif_results_univar_glm.xlsx')

```

```{r warning = F}
models <- formulas %>%       # add all formulas
  
  # iterate through each univariate formula
  map(                               
    .f = ~glm(                       # pass the formulas one-by-one to glm()
      formula = as.formula(.x),      # within glm(), the string formula is .x
      family = "binomial",           # specify type of glm (logistic)
      data = discharged)) %>%          # dataset
  
  # tidy up each of the glm regression outputs from above
  map(
    .f = ~tidy(
      .x, 
      exponentiate = TRUE,           # exponentiate 
      conf.int = TRUE)) %>%          # return confidence intervals
  
  # collapse the list of regression outputs in to one data frame
  bind_rows() %>% 
  # round all numeric columns
  mutate(across(where(is.numeric), round, digits = 2)) 
```

```{r}
library (gtsummary)
univ_tab <- discharged %>% 
  dplyr::select(all_of(x), `Ventilation`) %>% ## select variables of interest

  tbl_uvregression(                         ## produce univariate table
    method = glm,                           ## define regression want to run (generalised linear model)
    y = Ventilation,                        ## define outcome variable
    method.args = list(family = binomial),  ## define what type of glm want to run (logistic)
    exponentiate = TRUE                     ## exponentiate to produce odds ratios (rather than log odds)
  ) %>% 
bold_p()

## view univariate results table 
univ_tab
```

Graphical representation of univariant regression (color designate statistical significance)

```{r}

tmp1 %>% 
  
  #set order of levels to appear along y-axis
  mutate(`sign_.05` = fct_relevel(
    `sign_.05`, "yes", "no")) %>%
  
  # remove "intercept" row from plot
  filter(term != "(Intercept)") %>% 
  
   # remove "UNSIGNIFICANT" TERMS from plot
  filter (`sign_.05`!= "no") %>% 
  
  ## plot with variable on the y axis and estimate (OR) on the x axis
  ggplot(aes(x = estimate, y = term)) +
  
  ## show the estimate as a point
  geom_point(aes(color = `sign_.05`)) + 
  
  ## add in an error bar for the confidence intervals
  geom_errorbar(aes(xmin = conf.low, xmax = conf.high), width = 0.5) + 
  
  ## show where OR = 1 is for reference as a dashed line
  geom_vline(xintercept = 1, linetype = "dotted") +
  scale_x_continuous(limits = c(-1, 23))   # cut off extreme CI range
```

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!!!!!!!!!!!!НЕ ИСПРАВЛЯЛА !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Женя корреляции и логистическая регрессия

```{r}
library(ggplot2)
library(reshape2)
library(car)
names(data)



target_columns <- c(
  "Индекс_коморбидности_Чарлсона_1d",
  "CGS_балл",
  "Возраст",
  "FOUR_1d",
  "SOFA_1d",
  "WBC_1d",
  "NEU_#_1d",
  "LYM_#_1d",
  "PLT_1d",
  "Лактат_1d",
  "Креатинин_1d",
  "АЛТ_1d",
  "АСТ_1d",
  "Альбумин_1d",
  "Щелочная_фосфотаза_1d",
  "Мочевина_1d",
  "Мочевая_кислота_1d",
  "ГГТ_1d",
  "Калий_1d",
  "Натрий_1d",
  "АЧТВ_1d",
  "Хлорид_1d",
  "МНО_1d",
  "Фибриноген_1d",
  "ПВ_1d",
  "oneday_log",
  "8ohdg 1d log",
  "ИВЛ_д"
  
)


data_filtered <- data %>% select(all_of(target_columns))

names(data_filtered)
data_subset_numeric <- data_filtered %>%
  mutate(across(everything(), ~as.numeric(as.character(.)))) %>%
  select(where(is.numeric))  
correlation_matrix <- cor(data_subset_numeric, use = "complete.obs")
correlation_matrix


correlation_matrix_melted <- melt(correlation_matrix)

ggplot(correlation_matrix_melted, aes(Var1, Var2, fill = value)) +
  geom_tile() +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", midpoint = 0) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, color = "black"),  
    axis.text.y = element_text(color = "black")  
  ) +
  labs(title = "Матрица корреляций", x = "Переменные", y = "Переменные") -> corr_plot

#ggsave("C:/Users/enami/Downloads/correlation_matrix.png", plot = corr_plot, width = 10, #height = 8, bg = "white")

#data_filtered_num <- data %>%
# mutate(across(c("Возраст", "SOFA_1d", "Индекс_коморбидности_Чарлсона_1d", 
#                "CGS_балл", "oneday_log", "8ohdg 1d log"), as.numeric))

data$Исход <- as.numeric(data$Исход)
#drop_na(data)
data <- data[data$Исход %in% c(0, 1), ]
table(data$Исход)
str(data$Возраст)
data$SOFA_1d <- as.numeric(data$SOFA_1d)
data$Индекс_коморбидности_Чарлсона_1d <-as.numeric(data$Индекс_коморбидности_Чарлсона_1d)
data$CGS_балл<-as.numeric(data$CGS_балл)
data$oneday_log <-as.numeric(data$oneday_log)
data$`8ohdg 1d log` <-as.numeric(data$`8ohdg 1d log`)
data$Возраст <- as.numeric(data$Возраст)
model <- glm(Исход ~ Возраст + SOFA_1d + Индекс_коморбидности_Чарлсона_1d + 
               CGS_балл + oneday_log + `8ohdg 1d log`, 
             data = data, 
             family = binomial)
######LR ratio
data_clean <- na.omit(data[, c("Исход", "Возраст", "SOFA_1d", 
                               "Индекс_коморбидности_Чарлсона_1d", 
                               "CGS_балл", "oneday_log", "8ohdg 1d log")])

model_full <- glm(Исход ~ Возраст + SOFA_1d + Индекс_коморбидности_Чарлсона_1d + 
                  CGS_балл + oneday_log + `8ohdg 1d log`, 
                  data = data_clean, 
                  family = binomial)
model_reduced <- glm(Исход ~ Возраст, 
                     data = data_clean, 
                     family = binomial)
lrt_result <- anova(model_reduced, model_full, test = "Chi")
print(lrt_result)


library(car)
vif(model)
vif_results <- vif(model)
vif_results_df <- as.data.frame(vif_results)
vif_results_df$Variables <- rownames(vif_results_df)

summary(model)
logistic_results <- as.data.frame(summary(model)$coefficients)
logistic_results$Variables <- rownames(logistic_results)

###we should add OR and CI to report results #https://www.statology.org/how-to-report-logistic-regression-results/
logistic_results$OR <- exp(logistic_results$Estimate)
logistic_results$CI_lower <- exp(logistic_results$Estimate - 1.96 * logistic_results$`Std. Error`)
logistic_results$CI_upper <- exp(logistic_results$Estimate + 1.96 * logistic_results$`Std. Error`)

#install.packages("performance")
library(performance)
model_check <-check_model(model)
model_check
#ggsave("C:/Users/enami/Downloads/model_diagnostics_plot.png", plot = model_check)
#png("C:/Users/enami/Downloads/combined_model_diagnostics.png", width = 1200, height = 800)
#plot(model_check)  # Рисуем графики
#dev.off()
library(dagitty)

#####DAG ANALYSIS

#install.packages("ggdag")
#install.packages("dagitty")
library(ggdag)
library(dagitty)
dag <- dagitty("dag {
  Outcome <- Age
  Outcome <- SOFA_1d
  Outcome <- Charlson_Index_1d
  Outcome <- CGS_Score
  Outcome <- oneday_log
  Outcome <- 8hdg_1d_log
  
  Age -> SOFA_1d
  Age -> Charlson_Index_1d
  Age -> CGS_Score
  
  oneday_log -> SOFA_1d
  oneday_log -> Charlson_Index_1d
  8hdg_1d_log -> SOFA_1d
  8hdg_1d_log -> Charlson_Index_1d
}")


plot(dag)

dag_plot <- ggdag(dag) +
  geom_dag_node(color = "grey", size = 15, alpha = 0.7) +  
  geom_dag_text(aes(label = name), color = "black", size = 5, fontface = "bold") +  
  theme_dag() + 
  theme(
    legend.position = "none"
  ) +
  coord_fixed(ratio = 1.5)  

#ggsave("C:/Users/enami/Downloads/dag_plot.png", plot = dag_plot, width = 10, height = 8, #bg = "white")

model_dag <- glm(Исход ~ Возраст + SOFA_1d + Индекс_коморбидности_Чарлсона_1d + 
                   CGS_балл + oneday_log + `8ohdg 1d log` +
                   Возраст:SOFA_1d + Возраст:Индекс_коморбидности_Чарлсона_1d +
                   Возраст:CGS_балл + oneday_log:SOFA_1d + oneday_log:Индекс_коморбидности_Чарлсона_1d + 
                   `8ohdg 1d log`:SOFA_1d + `8ohdg 1d log`:Индекс_коморбидности_Чарлсона_1d, 
                 data = data, 
                 family = binomial)
model_dag

library(car)
vif_results_dag <- vif(model_dag)
vif_results_df_dag <- as.data.frame(vif_results_dag)
vif_results_df_dag$Variables <- rownames(vif_results_df_dag)
logistic_results_dag <- as.data.frame(summary(model_dag)$coefficients)
logistic_results_dag$Variables <- rownames(logistic_results_dag)

library(performance)
model_check_dag <- check_model(model_dag)
model_check_dag

#png("C:/Users/enami/Downloads/model_dag_diagnostics_plot.png", width = 1200, height = 800)
#plot(model_check)  
#dev.off()


library(ggdag)


list(
  VIF = vif_results_df_dag,
  Summary = logistic_results_dag,
  Diagnostics = model_check_dag
)

#####попробую вот эту штуку центрирование – вычитание собственной средней из всех ##интервальных предикторов; приводит к осмысленному значению константы
data_center <- data

data_center$Возраст <- scale(data$Возраст, center = TRUE, scale = FALSE)
data_center$SOFA_1d <- scale(data$SOFA_1d, center = TRUE, scale = FALSE)
data_center$Индекс_коморбидности_Чарлсона_1d <- scale(data$Индекс_коморбидности_Чарлсона_1d, center = TRUE, scale = FALSE)
data_center$CGS_балл <- scale(data$CGS_балл, center = TRUE, scale = FALSE)
data_center$oneday_log <- scale(data$oneday_log, center = TRUE, scale = FALSE)
data_center$`8ohdg 1d log` <- scale(data$`8ohdg 1d log`, center = TRUE, scale = FALSE)



model_dagcentr <- glm(Исход ~ Возраст + SOFA_1d + Индекс_коморбидности_Чарлсона_1d + 
                        CGS_балл + oneday_log + `8ohdg 1d log` +
                        Возраст:SOFA_1d + Возраст:Индекс_коморбидности_Чарлсона_1d +
                        Возраст:CGS_балл + oneday_log:SOFA_1d + oneday_log:Индекс_коморбидности_Чарлсона_1d + 
                        `8ohdg 1d log`:SOFA_1d + `8ohdg 1d log`:Индекс_коморбидности_Чарлсона_1d, 
                      data = data_center, 
                      family = binomial)
vif_results_dagcentr <- vif(model_dagcentr)
vif_results_df_dagcentr <- as.data.frame(vif_results_dagcentr)
vif_results_df_dagcentr$Variables <- rownames(vif_results_df_dagcentr)
vif_results_dagcentr #####круто центрирование сработало
logistic_results_dagcentr <- as.data.frame(summary(model_dagcentr)$coefficients)
logistic_results_dagcentr$Variables <- rownames(logistic_results_dagcentr)
logistic_results_dagcentr$OR <- exp(logistic_results_dagcentr$Estimate)
logistic_results_dagcentr$CI_lower <- exp(logistic_results_dagcentr$Estimate - 1.96 * logistic_results_dagcentr$`Std. Error`)
logistic_results_dagcentr$CI_upper <- exp(logistic_results_dagcentr$Estimate + 1.96 * logistic_results_dagcentr$`Std. Error`)

model_check_dagcentr <- check_model(model_dagcentr)
model_check_dagcentr

#png("C:/Users/enami/Downloads/model_dag_diagnostics_plot_scale.png", width = 1200, height #= 800)
#plot(model_check_dagcentr)  
#dev.off()
list(
  VIF = vif_results_df_dagcentr,
  Summary = logistic_results_dagcentr,
  Diagnostica = model_check_dagcentr
)

#####lets do the lrt again
data_clean_center <- na.omit(data_center[, c("Исход", "Возраст", "SOFA_1d", 
                               "Индекс_коморбидности_Чарлсона_1d", 
                               "CGS_балл", "oneday_log", "8ohdg 1d log")])
model_reduced_combo <- glm(Исход ~ Возраст + SOFA_1d + Индекс_коморбидности_Чарлсона_1d + 
                      CGS_балл + oneday_log + `8ohdg 1d log`, 
                    data = data_clean_center, 
                    family = binomial)

model_full_combo <- glm(Исход ~ Возраст + SOFA_1d + Индекс_коморбидности_Чарлсона_1d + 
                    CGS_балл + oneday_log + `8ohdg 1d log` +
                    Возраст:SOFA_1d + Возраст:Индекс_коморбидности_Чарлсона_1d +
                    Возраст:CGS_балл + oneday_log:SOFA_1d + oneday_log:Индекс_коморбидности_Чарлсона_1d + 
                    `8ohdg 1d log`:SOFA_1d + `8ohdg 1d log`:Индекс_коморбидности_Чарлсона_1d, 
                  data = data_clean_center, 
                  family = binomial)

lrt_result_combo <- anova(model_reduced_combo, model_full_combo, test = "Chi")

print(lrt_result_combo)
```

```{r}

```

### Лена унивариантный регрессия для всех переменных:

```{r}
# подготовить такой датасет лучше где -то выше (все переменные + включить группировку, обработанные полиморфизмы и NLR)

data_for_regres <- data %>% select (ID, Пол, Возраст, `анти-HCV`, `Hbs_AG`,  `ЛСМА`, Пневмония, Инфекция, diseaseGroup, CGS_балл, Индекс_коморбидности_Чарлсона_1d,`CIRS-G_сумма_1d`, `CIRS-G_систем_с_паталогиями_5d_36`, `CIRS-G_индекс_тяжести_1d`, FOUR_1d, SOFA_1d, внкДНК_1d, `8ohdg_1d`, WBC_1d,  `NEU_#_1d`, `NEU_%_1d`, `LYM_#_1d`, `LYM_%_1d`,  `PLT_1d`, `Лактат_1d`, Креатинин_1d, АЛТ_1d, АСТ_1d,  Альбумин_1d, Щелочная_фосфотаза_1d , Мочевая_кислота_1d, Мочевина_1d, ГГТ_1d, Калий_1d, Натрий_1d, Хлорид_1d,  АЧТВ_1d, МНО_1d, Фибриноген_1d, ПВ_1d, Уросептическая_инфекция, `Сепсис/септический_шок`, `Потребность_ИВЛ`, `Ухудшение_пневмонии_сутки`, `ИВЛ_д`, `Дни_ОРИТ_ФНКЦРР`, `Койко-дни_ФНКЦРР`, `Исход`, `Пневмония_возникшая_ФНКЦРР`, `oneday_log`, `8ohdg 1d log` ) %>% mutate(across(c(`CIRS-G_сумма_1d`, `CIRS-G_систем_с_паталогиями_5d_36`,`CIRS-G_индекс_тяжести_1d`,`FOUR_1d`, `WBC_1d`, `NEU_#_1d`, `NEU_%_1d`, `LYM_#_1d`, `LYM_%_1d`, `PLT_1d`, `Лактат_1d`, `Креатинин_1d` , `АЛТ_1d`, `АСТ_1d`, `Альбумин_1d`, `Щелочная_фосфотаза_1d`, `Мочевая_кислота_1d`, `Мочевина_1d` , `ГГТ_1d`, `Калий_1d`, `Натрий_1d`, `Хлорид_1d`, `АЧТВ_1d`, `МНО_1d`,`Фибриноген_1d`, ПВ_1d , `Ухудшение_пневмонии_сутки`, `ИВЛ_д`, `Дни_ОРИТ_ФНКЦРР`, `Койко-дни_ФНКЦРР`), as.numeric), across(c(Исход,`Потребность_ИВЛ`, factor)) 


hosp_poly_group <- left_join(data_for_regres, hospital_polymorf[,1:11], by = "ID" )

hosp_poly_group_nlr <- left_join(hosp_poly_group, NLR_data, by = "ID" ) %>% select (-ID)

#Получили полный датасет, включая новые переменные и отредактированные исходные c правильным типом
summary(hosp_poly_group_nlr)
colnames(hosp_poly_group_nlr)
```

Построим унивариантную регрессию для [ИСХОДА]

```{r}

library(broom)
library(openxlsx)

x = names(hosp_poly_group_nlr[,-48])
out <- unlist(lapply(48, function(n) combn(x, 1, FUN=function(row) paste0("Исход ~ ", paste0("`",row, "`",collapse = "+")))))
out

#regression coefficients
tmp1 = bind_rows(lapply(out, function(frml) {
 a = tidy(glm(frml, data= hosp_poly_group_nlr, family = binomial))
 a$frml = frml
 return(a)
}))

tmp1

#regression results R2, AIC, BIC
tmp2 = bind_rows(lapply(out, function(frml) {
 a = glance(lm(frml, data=hosp_poly_group_nlr))
 a$frml = frml
 return(a)
}))
tmp2

signif_estim_univar <- as_data_frame(tmp1) %>% filter (`p.value` <= 0.05)
write.xlsx(signif_estim_univar, 'signif_estim_univar_glm.xlsx')

#signif_results_univar<- as_data_frame(tmp2)
#write.xlsx(signif_estim_univar, 'signif_results_univar_glm.xlsx')
```
###  Лена Регрессионный анализ для поиска ассоциации:

Проверить гипотезы о взаимодействии (для всей выборки и в стратах)
○ Полиморфизм гена TLR9 (рецептор) и внеклеточной ДНК 
○ Полиморфизм гена AQP4 и лейкоцитарного индекса NLR

##### Cвязь полиморфизма гена TLR9 (рецептор) и внеклеточной ДНК 

In nuclear and mitochondrial DNA, 8-hydroxy-2' -deoxyguanosine (8-OHdG) or 8-oxo-7,8-dihydro-2' -deoxyguanosine (8-oxodG) is one of the predominant forms of free radical-induced oxidative lesions, and has therefore been widely used as a biomarker for oxidative stress and carcinogenesis. Studies showed that urinary 8-OHdG is a good biomarker for risk assessment of various cancers and degenerative diseases. 

```{r}

disease_group_id <- data %>% select (ID, Возраст, SOFA_1d,  diseaseGroup, oneday_log, `8ohdg 1d log`, CGS_балл,`Индекс_коморбидности_Чарлсона_1d`, `Койко-дни_ФНКЦРР`)



#  `ИСХОД` БЕЗ ГРУППИРОВКИ
fit_tlr <-   glm (`Исход` ~ TLR9_rs352162 :oneday_log + TLR9_rs187084 + `8ohdg 1d log`, data = hosp_poly_group_nlr, family = binomial)
summary(fit_tlr)
# Полиморфизм/ ДНК нет значимой ассоциации предикторов с исходом (выписка, смерть)

#  `ИСХОД` C ГРУППИРОВКОЙ
fit_tlr_gr <-   glm (`Исход` ~ TLR9_rs352162 + TLR9_rs187084 + oneday_log + `8ohdg 1d log` + diseaseGroup , data = hosp_poly_group_nlr, family = binomial)
summary(fit_tlr_gr)
# Полиморфизм/ ДНК не ассоциированы...Последствие ЧМТ ухудшает прогноз (ур знач 0,05)

#  `ПОТРЕБНОСТЬ В ИВЛ` C ГРУППИРОВКОЙ
fit_tlr_1 <-   glm (`Потребность_ИВЛ` ~ TLR9_rs352162 + TLR9_rs187084 + oneday_log + `8ohdg 1d log` + diseaseGroup , data = hosp_poly_group_nlr, family = binomial)
summary(fit_tlr_1)
# Полиморфизм/ ДНК не ассоциированы...Гипоксия ассоциирована с увеличением вероятности потребности в ИВЛ (ур знач 0,05)

#  `КОЙКО-ДНИ ФНКЦРР` C ГРУППИРОВКОЙ
fit_tlr_2 <-   lm (`Койко-дни_ФНКЦРР` ~ TLR9_rs352162 + TLR9_rs187084 + oneday_log + `8ohdg 1d log` + diseaseGroup, data = hosp_poly_group_nlr)
summary(fit_tlr_2)
# Полиморфизм/ ДНК не ассоциированы... Группа Последствие ЧМТ - положительная ассоциация с кол-вом койко-дней и отрицательная для Группы Новообразования


```

##### Cвязь полиморфизма гена AQP4 (рецептор) и уровня показателя NLR

```{r}

```

```{r}
#  `ИСХОД` БЕЗ ГРУППИРОВКИ
#  !!!!!! огромные std errors ??????

#  `ИСХОД` C ГРУППИРОВКОЙ
fit_nlr_gr <-   glm (`Исход` ~  AQP4_rs1058427 + NLR_1d + diseaseGroup , data = hosp_poly_group_nlr, family = binomial)
summary(fit_nlr_gr)
# NLR_1 ассоциирован с исходом (эффект незначителен?), страта не ассоциирована

#  `ПОТРЕБНОСТЬ В ИВЛ` C ГРУППИРОВКОЙ
fit_nlr_gr1 <-   glm (`Потребность_ИВЛ` ~  AQP4_rs1058427 + NLR_1d + diseaseGroup , data = hosp_poly_group_nlr, family = binomial)
summary(fit_nlr_gr1)
## NLR_1 ассоциирован с исходом (эффект незначителен?), новообразование?

#  `КОЙКО-ДНИ ФНКЦРР` C ГРУППИРОВКОЙ
fit_nlr_2 <-   lm (`Койко-дни_ФНКЦРР` ~ AQP4_rs1058427 + NLR_1d + diseaseGroup, data = hosp_poly_group_nlr)
summary(fit_nlr_2)
# Полиморфизм/ NLR не ассоциированы... Последствия ЧМТ?
```


