---
title: "Identification of predictors of outcome in patients with consequences of severe brain injures"
output: html_document
author: "Elena Iurlova"
date: "2024-12-07"
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE)
library(MASS)
library(stats)
library(ggthemes)
library (gtsummary)
library(factoextra)
library(FactoMineR)
library(ggbiplot) 
library(plotly)
library(forcats)
library(corrplot)
library(flextable)
library(readxl)
library(lubridate)
library(broom)
library(openxlsx)
library(anytime)
library(SNPassoc)
library(tidyverse)
```

## Dataset correction preparation

Dates correction

```{r cars}
data <- read_csv2("data.csv")

data$`Дата перевода в ФНКРР`[191] <- "09.11.2018"
data$`Дата мозговой катастрофы`[8] <- "01.01.2014"
data$`Дата мозговой катастрофы`[169] <- "12.09.2017"
data$`Дата мозговой катастрофы`[321] <- "01.01.2014"

data <- data %>% mutate(`Дата мозговой катастрофы` = dmy(`Дата мозговой катастрофы`),
                     `Дата перевода в ФНКРР` = dmy(substr(`Дата перевода в ФНКРР`, start=1, stop=8)),
                     `Дата выписки из ФНКЦ РР` = dmy(`Дата выписки из ФНКЦ РР`)
                     )

```

DNA data preparation. Determined coefficients between 2 kits and multiplied the values of rows 1:174 by the coefficient. Log transformation was applied to total extracellular DNA ammount and to 8-Hydroxyguanosine measurements.

Note: In nuclear and mitochondrial DNA, 8-hydroxy-2' -deoxyguanosine (8-OHdG) or 8-oxo-7,8-dihydro-2' -deoxyguanosine (8-oxodG) is one of the predominant forms of free radical-induced oxidative lesions, and has therefore been widely used as a biomarker for oxidative stress and carcinogenesis. Studies showed that urinary 8-OHdG is a good biomarker for risk assessment of various cancers and degenerative diseases. 

```{r}
######## Replace н/д and normalize variable `внк ДНК`
data <- data %>%
  mutate(across(everything(), ~na_if(as.character(.), "н/д")))

##data <- data %>%
  ##mutate(`Дата мозговой катастрофы` = dmy(`Дата мозговой катастрофы`))
target_columns <- c("внкДНК, нг/мл 1d", "8ohdg на миллион оснований 1d",
                    "внкДНК, нг/мл 5d", "8ohdg на миллион оснований 5d")


data <- data %>%
  mutate(across(all_of(target_columns), as.numeric))


data <- data %>%
  mutate(
    oneday_log = if_else(`внкДНК, нг/мл 1d` > 0, log(`внкДНК, нг/мл 1d`), NA_real_),
    fiveday_log = if_else(`внкДНК, нг/мл 5d` > 0, log(`внкДНК, нг/мл 5d`), NA_real_)
  )


data1 <- data %>% slice(1:173)
data2 <- data %>% slice(220:522)

mean1 <- mean(data1$oneday_log, na.rm = TRUE)
mean2 <- mean(data2$oneday_log, na.rm = TRUE)

coeff1 <- mean2 / mean1 ##determine correction coefficients for data for 1 day DNA measurements

data <- data %>%
  mutate(oneday_log = if_else(row_number() <= 173, 
                              oneday_log * coeff1, 
                              oneday_log))

data1 <- data %>% slice(1:173)
data2 <- data %>% slice(220:522)

mean1 <- mean(data1$fiveday_log, na.rm = TRUE)
mean2 <- mean(data2$fiveday_log, na.rm = TRUE)

coeff2 <- mean2 / mean1 ##determine correction coefficients for data for 5 day DNA measurements

data <- data %>%
  mutate(fiveday_log = if_else(row_number() <= 173, 
                               fiveday_log * coeff2, 
                               fiveday_log))
###### 8ohg 
data <- data %>%
  mutate(
    `8ohdg 1d log` = if_else(`8ohdg на миллион оснований 1d` > 0, log(`8ohdg на миллион оснований 1d`), NA_real_),
    `8ohdg 5d log` = if_else(`8ohdg на миллион оснований 5d` > 0, log(`8ohdg на миллион оснований 5d`), NA_real_)
  )

data1 <- data %>% slice(1:173)
data2 <- data %>% slice(220:522)

mean1_ohdg1 <- mean(data1$`8ohdg 1d log`, na.rm = TRUE)
mean2_ohdg1 <- mean(data2$`8ohdg 1d log`, na.rm = TRUE)

coeff_ohdg1 <- mean2_ohdg1 / mean1_ohdg1 ##determine correction coefficients for data for 5 day DNA measurements

data <- data %>%
  mutate(`8ohdg на миллион оснований 1d лог` = if_else(row_number() <= 173, 
                                                      `8ohdg 1d log` * coeff_ohdg1, 
                                                      `8ohdg 1d log`))
mean1_ohdg5 <- mean(data1$`8ohdg 5d log`, na.rm = TRUE)
mean2_ohdg5 <- mean(data2$`8ohdg 5d log`, na.rm = TRUE)

coeff_ohdg5 <- mean2_ohdg5 / mean1_ohdg5

data <- data %>%
  mutate(`8ohdg 5d лог` = if_else(row_number() <= 173, 
                                                      `8ohdg 5d log` * coeff_ohdg5, 
                                                      `8ohdg 5d log`))
data_e <- data # clone dataset
```

 List of primary variables (according to medical basis) : 
  Номер, ID,  Пол, HTR1A_rs6295, AQP4_rs1058427 , BDNF_rs6265, OLR1_rs11053646, NRF2_rs6726395, AGTR1_rs275651, AQP5_rs3759129, TLR9_rs352162, TLR9_rs187084, AQP_rs3736309, Возраст, анти-HCV, Hbs_AG,  ЛСМА, Основной_диагноз_код_МКБ10, Диагноз, Пневмония, Инфекция, CGS_балл, Индекс_коморбидности_Чарлсона_1d, CIRS-G_сумма_1d, CIRS-G_систем_с_паталогиями_1, CIRS-G_индекс_тяжести_1d, FOUR_1d, SOFA_1d, внкДНК_1d, 8ohdg_1d, WBC_1d,  NEU_#_1d, NEU_%_1d, LYM_#_1d, LYM_%_1d,  PLT_1d, Лактат_1d, Креатинин_1d, АЛТ_1d, АСТ_1d,  Альбумин_1d, Щелочная_фосфотаза_1d , Мочевая_кислота_1d, Мочевина_1d, ГГТ_1d, Калий_1d, Натрий_1d, Хлорид_1d,  АЧТВ_1d, МНО_1d, Фибриноген_1d, ПВ_1d, Уросептическая_инфекция, Сепсис/септический_шок, Потребность_ИВЛ, Ухудшение_пневмонии_сутки, ИВЛ_д, Дни_ОРИТ_ФНКЦРР, Койко-дни_ФНКЦРР, Исход, Пневмония_возникшая_ФНКЦРР 

#### Data rename and correction

```{r}
data_e <- data_e %>% 
  rename ("Num" = "№ истории", "Part" = "Часть", "Sex" = "Пол", "Num_ref" = "Номер Настя", "Age" = "Возраст", "HTR1A_rs6295" = "HTR1A  rs6295","AQP4_rs1058427"  = "AQP4 rs1058427", "BDNF_rs6265" = "BDNF rs6265", "OLR1_rs11053646" = "OLR1 rs11053646", "NRF2_rs6726395" = "NRF2 rs6726395", "AGTR1_rs275651" = "AGTR1  rs275651",  "AQP5_rs3759129" = "AQP5 rs3759129", "TLR9_rs352162" = "TLR9 rs352162", "TLR9_rs187084" = "TLR 9 rs187084", "AQP5_rs3736309" = "AQP rs3736309",  "anti-HCV" = "анти HCV 2-нд, 1 - да, 0 - нет", "Hbs_AG" = "Hbs AG 0-не,2-нд, 1 - да", "LMCA" = "ЛСМА", "Date_brain_damage" = "Дата мозговой катастрофы" , "Main_diagnosis_code" = "Основной диагноз код МКБ10", "Diagnosis" = "Диагноз целиком", "Transfer_date_FSCCRR" = "Дата перевода в ФНКРР", "Pneumonia" = "Поступил в ФНКЦ с пневмонией", "Infection"  = "Поступил в ФКНЦ с инфекцией", "T_admission"  = "t° при поступлении", "CGS" = "CGS, балл", "CCI_1d"  = "Индекс коморбидности Чарлсона (баллы) 1d",  "CCI_5d" = "Индекс коморбидности Чарлсона (баллы) 5d",
"CCI_21d" = "Индекс коморбидности Чарлсона (баллы) 21d",
"CCI_out" = "Индекс коморбидности Чарлсона (баллы)  out",
"CIRS-G_sum_1d"  = "CIRS-G сумма, 1d",
"CIRS-G_pathology_1d" = "CIRS-G систем с паталогиями, 1d" ,
"CIRS-G_severity_1d" = "CIRS-G индекс тяжести, 1d",
"CIRS-G_sum_5d_35"  = "CIRS-G сумма, 5d...35",
"CIRS-G_pathology_5d_36"= "CIRS-G систем с паталогиями, 5d...36",
"CIRS-G_severity_5d_37" = "CIRS-G индекс тяжести, 5d...37",
"CIRS-G_sum_5d_38" = "CIRS-G сумма, 5d...38",
"CIRS-G_pathology__5d_39"  = "CIRS-G систем с паталогиями, 5d...39",
"CIRS-G_severity_5d_40" = "CIRS-G индекс тяжести, 5d...40",
"CIRS-G_sum_out" = "CIRS-G сумма, out",
"CIRS-G_pathology_out" = "CIRS-G систем с паталогиями, out",
"CIRS-G_severity_out" = "CIRS-G индекс тяжести, out",
"exDNA_1d" = "внкДНК, нг/мл 1d",
"8ohdg_1d" = "8ohdg на миллион оснований 1d",
"exDNA_5d" = "внкДНК, нг/мл 5d",
"8ohdg_5d" = "8ohdg на миллион оснований 5d",
"Uroseptic_infection" = "Уросептическая инфекция в любой момент времени",
"Sepsis/septic_shock" = "Сепсис/септический шок в любой момент времни",
"Ventilation" = "Потребность в ИВЛ за время госпитализации в любой момент времени",
"Ухудшение пневмонии_сутки" = "Ухудшение исходной пневмонии, сутки",
"Ventilation_d" = "ИВЛ (дни)",
"Discharge_date_FSCCRR" = "Дата выписки из ФНКЦ РР",
"ICU_FSCCRR_d" = "Дней в ОРИТ ФНКЦ РР",
"FSCCRR_beddays" = "Койко-дней в ФНКЦ РР",
"Outcome" = "Исход, 0 -выписан, 1 умер, 2 - цензура",
"Pneumonia_at_FSCCRR" = "Пневмония возникшая в ФНКЦ РР",
 "Pneumonia_deterioration_d" = "Ухудшение исходной пневмонии, сутки") %>% 
  rename_with(~ gsub("Лактат","Lactate", .x)) %>% 
  rename_with(~ gsub("Креатинин","Creatinine", .x)) %>% 
  rename_with(~ gsub("АЛТ","ALT", .x)) %>% 
  rename_with(~ gsub("АСТ","AST", .x)) %>% 
  rename_with(~ gsub("Альбумин","Albumin", .x)) %>% 
  rename_with(~ gsub("Щелочная фосфотаза","Alcaline_phosphatase", .x)) %>% 
  rename_with(~ gsub("Мочевая кислота","Uric_acid", .x)) %>% 
  rename_with(~ gsub("Мочевина","Urea", .x)) %>% 
  rename_with(~ gsub("ГГТ","GGT", .x)) %>% 
  rename_with(~ gsub("Натрий","Sodium", .x)) %>% 
  rename_with(~ gsub("Калий","Potassium", .x)) %>% 
  rename_with(~ gsub("Хлорид","Chloride", .x)) %>% 
  rename_with(~ gsub("АЧТВ","APTT", .x)) %>% 
  rename_with(~ gsub("МНО","INR", .x)) %>% 
  rename_with(~ gsub("Фибриноген","Fibrinogen", .x)) %>% 
  rename_with(~ gsub("ПВ","PT", .x)) %>% 
  rename_at(vars(44 : 148), ~str_replace_all(., "\\s+", "_"))  %>% 
  mutate(across(`Age`, as.integer)) 
  
data_e <- data_e[ ,-c(52, 54, 163, 164)] ## delete initial dna- variables  

```

Calculate term before transfer to FSCCRR (`Before_FSCCRR_d`) and clean outcome variables (н/д replace for NA, Outcome 2 - for NA), also check and correct letter case for sex designation

Create new variable NLR_1d:

```{r }

data_e <- data_e %>% mutate (`Before_FSCCRR_d` = as.duration(interval(`Date_brain_damage`, `Transfer_date_FSCCRR`)) / as.duration(days(1)))%>% 
                     mutate(`Ventilation` = ifelse(`Ventilation` =="н/д", NA, `Ventilation`), `Outcome` = ifelse( `Outcome` == 2, NA, `Outcome`),
   `Sex` = case_when(`Sex` %in% c('ж', 'Ж') ~ 'F', `Sex`  %in% c('м', 'М') ~ 'M', .default = `Sex`),
   `Pneumonia` = case_when(`Pneumonia` == 1 ~ "Yes", `Pneumonia` == 0 ~ "No", TRUE ~ NA),
   `Infection` = case_when(grepl("1",`Infection` ) ~ "Yes", `Infection` == 0 ~ "No", TRUE ~ NA),
   `Sepsis/septic_shock` = case_when(grepl("1",`Sepsis/septic_shock` ) ~ "Yes", `Sepsis/septic_shock` == 0 ~ "No", TRUE ~ NA),
   `Uroseptic_infection` = case_when(grepl("1",`Uroseptic_infection` ) ~ "Yes", `Uroseptic_infection` == 0 ~ "No", TRUE ~ NA),
   `Pneumonia_at_FSCCRR` = case_when(grepl("1",`Pneumonia_at_FSCCRR` ) ~ "Yes", `Pneumonia_at_FSCCRR` == 0 ~ "No", TRUE ~ NA),
   `Hbs_AG`= case_when(`Hbs_AG` == 1 ~ "Yes", `Hbs_AG` == 0 ~ "No", TRUE ~ NA),
   `Ventilation`= case_when(`Ventilation` == 1 ~ "Yes", `Ventilation` == 0 ~ "No", TRUE ~ NA)) %>% 
                   mutate (`NLR_1d` = as.numeric(`NEU_#_1d`) / as.numeric(`LYM_#_1d`))

```

Note about NLR:
NLR reflects online dynamic relationship between innate (neutrophils) and adaptive cellular immune response (lymphocytes) during illness and various pathological states. NLR is influenced by many conditions including age, rice, medication, chronic disease like coronary heart disease, stroke, diabetes, obesity, psychiatric diagnosis, cancer of solid organs, anemia and stress. A normal range of NLR is between 1-2, the values higher than 3.0 and below 0.7 in adults are pathological. NLR in a grey zone between 2.3-3.0 may serve as early warning of pathological state or process such like cancer, atherosclerosis, infection, inflammation, psychiatric disorders and stress. 


### SNP data preparation and analysis

SNP : "HTR1A_rs6295" "AQP4_rs1058427" "BDNF_rs6265"
"OLR1_rs11053646" "NRF2_rs6726395" "AGTR1_rs275651" "AQP5_rs3759129"
"TLR9_rs352162" "TLR9_rs187084" "AQP5_rs3736309"


Generate separate dataset with corrected SNP, outcomes (`Ventilation`, `Outcome`,`FSCCRR_beddays`), DNA variables ( `oneday_log`, `8ohdg 1d log`) and NLR variable.

```{r}

hospital_polymorf_e <- data_e %>% 
                    mutate (`HTR1A_rs6295` = ifelse (`HTR1A_rs6295` == 'С/С', 'C/C', `HTR1A_rs6295`),
                            `AQP4_rs1058427` = ifelse (`AQP4_rs1058427` == 'C/A'| `AQP4_rs1058427` == 'С/A', 'C/A', `AQP4_rs1058427`),
                             `OLR1_rs11053646` = ifelse (`OLR1_rs11053646` == 'C/G', 'G/C', `OLR1_rs11053646`),
                             `AQP5_rs3759129` = ifelse (`AQP5_rs3759129` == 'С/C', 'C/C',`AQP5_rs3759129`),
                             `TLR9_rs352162` = case_when(`TLR9_rs352162` == 'С/С' | `TLR9_rs352162` == 'С/C' ~ 'C/C',
                             `TLR9_rs352162` == 'С/T' ~ 'C/T',.default = `TLR9_rs352162` ),
                             `TLR9_rs187084` = case_when(`TLR9_rs187084` == 'C/T' | `TLR9_rs187084` == 'С/T' ~ 'C/T', .default = `TLR9_rs187084` ),
                             `AQP5_rs3736309` = case_when(`AQP5_rs3736309` == 'C/T' | `AQP5_rs3736309` == 'С/T' ~ 'C/T',
                                                        .default = `AQP5_rs3736309` ))  %>% 
                   dplyr::select(ID, HTR1A_rs6295, AQP4_rs1058427, BDNF_rs6265, OLR1_rs11053646, NRF2_rs6726395, AGTR1_rs275651, AQP5_rs3759129, TLR9_rs352162, TLR9_rs187084, AQP5_rs3736309, `Ventilation`, `Outcome`, `oneday_log`, `8ohdg 1d log`, `NLR_1d`, `FSCCRR_beddays`) %>% 
                   mutate (across(c(HTR1A_rs6295, AQP4_rs1058427, BDNF_rs6265, OLR1_rs11053646, NRF2_rs6726395, AGTR1_rs275651, AQP5_rs3759129, TLR9_rs352162, TLR9_rs187084, AQP5_rs3736309), factor), across(`FSCCRR_beddays`, as.numeric)) 

```

SNP distribution graph

```{r fig.height=11, fig.width=11}
hospital_polymorf_long <- hospital_polymorf_e %>%
  dplyr::select (ID, `HTR1A_rs6295`, `AQP4_rs1058427`, `BDNF_rs6265`, `OLR1_rs11053646`, `NRF2_rs6726395`, `AGTR1_rs275651`, `AQP5_rs3759129`, `TLR9_rs352162`, `TLR9_rs187084`, `AQP5_rs3736309`) %>%
  pivot_longer(!ID, names_to = "Position", values_to = "SNP")

theme_cust <- theme(
    axis.text.x = element_text(size = 14, face = 'bold'),
    axis.text.y = element_text(size = 14),
    axis.title = element_text(size = 14),
    plot.title = element_text(hjust=0.5),
    legend.title = element_text(size = 14),
    legend.text = element_text(size = 14),
    panel.background = element_rect(fill = "white", color = 'black'),
    plot.background = element_rect(fill = "white"),
    ##panel.grid.major = element_line(color = "gray90"),
    ##panel.grid.minor = element_line(color = "lightgray"),
    strip.text = element_text(size = 14)
  )

ggplot (hospital_polymorf_long) +
  geom_bar (aes(SNP, fill = SNP)) +
  facet_wrap (~Position, nrow = 5, axes = "all", axis.labels = "all_x") +
  theme_base() +
  scale_fill_brewer(palette = "RdYlGn") +
  theme_cust +
  labs (title = ' SNP distribution')
```


**SNP descriptive analysis**

```{r}
library(SNPassoc)
## specify columns contain SNP, filter only discharged patients
polymorf.s <- setupSNP (data=hospital_polymorf_e %>% filter(Outcome == "0"), colSNPs=2:11, sep="/")
## got new class snp

head(polymorf.s$TLR9_rs352162)
summary(polymorf.s$TLR9_rs352162)
## which shows the genotype and allele frequencies for a given SNP, testing for Hardy-Weinberg equilibrium (HWE).
```

```{r}
 #We can  visualize the results in a plot by
plot(polymorf.s$TLR9_rs352162)
plot(polymorf.s$TLR9_rs352162, type=pie)
```

Genotyping of SNPs needs to pass quality control measures. Aside from technical details that need to be considered for filtering SNPs with low quality, genotype calling error can be detected by a HWE test. The test compares the observed genotype frequencies with those expected under random mating, which follows when the SNPs are in the absence of selection, mutation, genetic drift, or other forces. Therefore, HWE must be checked only in controls. There are several tests described in the literature to verify HWE. In SNPassoc HWE is tested for all the bi-allelic SNP markers using a fast exact test Wigginton, Cutler, and Abecasis (2005) implemented in the tableHWE function.

```{r}
##results across all snp
##prepare flextable (summary can't be converted directly)
cbind(colnames(polymorf.s)[7:16] %>% enframe(), summary(polymorf.s, print=FALSE)) %>% 
                                     rename ("ID"= "name", "SNP" = "value") %>% 
                                     mutate(across(where(is.numeric), round, digits = 2)) %>%
                                         flextable()

```

Showing the SNP labels with minor/major allele format, the major allele
frequency the HWE test and the percentage of missing genotypes. Missing
values can be further explored plotting with:

```{r}
plotMissing(polymorf.s, print.labels.SNPs = FALSE)
```
Black squares shows missing genotype information of dataset.
This plot can be used to inspect if missing values appear randomly across individuals and SNPs. In our case, we can see that the missing pattern may be considered not random. Group of individuals should be further checked for possible problems with genotyping.
MNCR?


**Hardy-Weinberg equilibrium**

Genotyping of SNPs needs to pass quality control measures. Aside from technical details that need to be considered for filtering SNPs with low quality. The test compares the observed genotype frequencies with those expected under random mating, which follows when the SNPs are in the absence of selection, mutation, genetic drift, or other forces. Therefore, HWE must be checked only in controls. There are several tests described in the literature to verify HWE. In SNPassoc HWE is tested for all the bi-allelic SNP markers using a fast exact test Wigginton, Cutler, and Abecasis (2005) implemented in the tableHWE function.


```{r}
hwe <- tableHWE(polymorf.s)
hwe
```
We observe that TLR9_rs352162 and TLR9_rs187084  SNPs in the dataset is not under HWE since their P-values rejecting the HWE hypothesis (null hypothesis) are less than 0.05. However, when tested in control samples only (appreciated attitude), by stratifying by cases and controls.

Set Ventillation demand as a Outcome for further HWE analysis.

 One is interested in keeping those SNPs that **do not reject the null hypothesis**. 

NOTE: As several SNPs are tested, multiple comparisons must be considered. In this particular setting, a threshold of 0.001 is normally considered. As a quality control measure, it is not necessary to be as conservative as in those situations where false discovery rates need to be controlled.


```{r}
hwe3 <- tableHWE(polymorf.s, `Ventilation`)
##prepare flextable
cbind (colnames(polymorf.s)[7:16] %>% enframe(), hwe3 %>% as_tibble()) %>%
                                 rename ("ID"= "name", "SNP" = "value") %>%
                               mutate(across(where(is.numeric), round, digits = 2)) %>%
                                         flextable()

##SNPs is HWE in the whole sample but not controls
snpNHWE <- hwe3[,1]>0.05 & hwe3[,2]<0.05
rownames(hwe3)[snpNHWE]
```

The decision was to assign all SNP to further analysis (all SNP is under HWE).

```{r}
##SNPs that do not pass the HWE test must be removed form further analyses. We can recall setupSNP and indicate the columns of the SNPs to be kept

##snps.ok <- rownames(hwe2)[hwe2[,2]>=0.001]
##pos <- which(colnames(hospital_polymorf_e)%in%snps.ok, useNames = FALSE)
##polymorph.s <- setupSNP(hospital_polymorf_e, pos, sep="/")
```


**SNP association analysis**
**Outcome analysis**

We are interested in finding SNPs associated with outcome status. The association analysis for all genetic models is performed by the function association that regresses Outcome on the variable SNP from the dataset that already contains the SNP variables of class snp.

For single SNP:

```{r}
association(`Ventilation` ~ TLR9_rs187084 , data= polymorf.s)
```

Univariate testing for all SNP.
The P-values obtained from massive univariate analyses are visualized with the generic plot function

```{r}
ans <- WGassociation(`Ventilation`, data=polymorf.s)
ans
plot(ans) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 20, vjust = 0.5, hjust=1))
```
Manhattan-type plots for different genetic models. P-values in -log10 scale to assess the association between case-control status and SNPs.
This produces a Manhattan plot of the -log10(P-values) for all the SNPs over all models. It shows the nominal level of significance and the Bonferroni level, which is the level corrected by the multiple testing across all SNPs. The overall hypothesis of massive univariate association tests is whether there is any SNP that is significantly associated with the phenotype. As multiple SNPs are tested, the probability of finding at least one significant finding increases if we do not lower the significance threshold. The Bonferroni correction lowers the threshold by the number of SNPs tested (0.005=0.05/10). In the Manhattan plot of our analysis, we see that no SNP is significant at the Bonferroni level, and therefore there is no SNP that is significantly associated with Outcome.

Testing hypthesis about association between NLR level and

```{r}
association(`NLR_1d` ~ `AQP4_rs1058427` , data= polymorf.s)
```


```{r}
ans_nlr <- WGassociation(`NLR_1d`, data=polymorf.s)
ans_nlr
plot(ans_nlr) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 20, vjust = 0.5, hjust=1))
```

```{r}
ans_dna <- WGassociation(`FSCCRR_beddays`, data=polymorf.s)
ans_dna
plot(ans_dna) +
   theme_minimal() +
  theme(axis.text.x = element_text(angle = 40, vjust = 0.5, hjust=0.7, size = 14))
```

### Create new variable "deciase group" for classification of diagnoses codes

```{r}
data_e <- data_e %>% mutate(
    `Main_diagnosis_code` = case_when(
    `Main_diagnosis_code` == "I 69.1" ~ "I69.1",
    `Main_diagnosis_code` == "I 69.3" ~ "I69.3",
    `Main_diagnosis_code` == "G.93.1" ~ "G93.1",
    `Main_diagnosis_code` == "I.69." ~ "I69",
    `Main_diagnosis_code` == "I.69.3" ~ "I69.3",
    `Main_diagnosis_code` == "Т.94." ~ "Т94",
    `Main_diagnosis_code` == "I 69.0" ~ "I69.0",
    `Main_diagnosis_code` == "Т90." ~ "Т90",
    `Main_diagnosis_code` == "I.21.4" ~ "I21.4",
    `Main_diagnosis_code` == "I.69.1" ~ "I69.1",
    `Main_diagnosis_code` == "Т 90.5" ~ "Т90.5",
    `Main_diagnosis_code` == "D33.0 D" ~ "D33.0",
    `Main_diagnosis_code` == "M86.1 О" ~ "M86.1",
    `Main_diagnosis_code` == "G93.1 G" ~ "G93.1",
    `Main_diagnosis_code` == "67.9 Це" ~ "I67.9",
    `Main_diagnosis_code` == "G93.1 А" ~ "G93.1",
    `Main_diagnosis_code` == "G09 ос" ~ "G09",
    `Main_diagnosis_code` == "I69.0 1" ~ "I69.0",
    `Main_diagnosis_code` == "G. 93.1" ~ "G93.1",
    `Main_diagnosis_code` == "26" ~ "I69.3",   ##!!!!
    `Main_diagnosis_code` == "G93..4" ~ "G93.4",
    `Main_diagnosis_code` == "169.3"	~ "I69.3",
    `Main_diagnosis_code` == "169.1"	~ "I69.1",
    `Main_diagnosis_code` == "160.0"	~ "I60.0",
    `Main_diagnosis_code` == "169.4" ~ "I69.4",
    `Main_diagnosis_code` == "T90,5" ~ "T90.5",
    `Main_diagnosis_code` == "T.94.0" ~ "T94.0",
    TRUE ~ `Main_diagnosis_code`
  ),
  `Main_diagnosis_code` = str_replace_all(`Main_diagnosis_code`, "С", "C"),
  `Main_diagnosis_code` = str_replace_all(`Main_diagnosis_code`, "Т", "T"),
  `Main_diagnosis_code` = str_replace_all(`Main_diagnosis_code`, ";", ""),
  `Main_diagnosis_code` = str_replace_all(`Main_diagnosis_code`, ", ", ";"),
  `Main_diagnosis_code` = str_replace_all(`Main_diagnosis_code`, "    ", " "),
  `Main_diagnosis_code` = str_replace_all(`Main_diagnosis_code`, " ", ";"),
)
```

```{r}
data_e %>% summarise(
  diseaseGroup = case_when(
    `Main_diagnosis_code` %in% c("I63", "I69.3", "I69.4", "T82.8") ~ "Ischemic stroke",
    `Main_diagnosis_code` %in% c("Q28", "I60", "I72", "I69.2", "I69.1", "I69.0", "I61") ~ "Hemorrhagic stroke",
    `Main_diagnosis_code` %in% c("S06", "T94", "T94") ~ "TBI",
    `Main_diagnosis_code` %in% c("G93.4", "G93.1", "G94.3", "I46", "U07.1", "U09.9", "G96", "I21", "I23.1") ~ "Hypoxia",
    `Main_diagnosis_code` %in% c("С71", "D33", "D32", "D43", "I72") ~ "Tumor",
    TRUE ~ "OTHER"
  ), `Main_diagnosis_code`) %>% filter(diseaseGroup == "OTHER") %>% unique()
```


```{r}
## Classify diagnisis into groups
data_e <- data_e %>% mutate(
  diseaseGroup = case_when(
    `Main_diagnosis_code` %in% c("I63", "I69.3", "I69.4", "T82.8", "I69.3;C34.1", "I63.5", "I63.3", "I63.2", "I63.4", "I63.9") ~ "Ischemic stroke",
    `Main_diagnosis_code` %in% c("Q28", "I60", "I72", "I69.2", "I69.1", "I69.0", "I61", "I60.2", "I61.6", "I61.5", "I61.3", "I60.3", "I60.6", "I60.0") ~ "Hemorrhagic stroke",
    `Main_diagnosis_code` %in% c("S06", "T94", "T90", "T90.4", "T94.0", "T94", "T90.1", "T90", "T90.2", "T90.5", "T90.8", "S06.6", "S06.31", "S06.2;T88.8", "T94.1") ~ "TBI",
    `Main_diagnosis_code` %in% c("G93.4", "G93.1", "G94.3", "I46", "U07.1", "U09.9", "G96", "I21", "I23.1", "I46;G93.1", "I21.4", "G94.3;G82.4") ~ "Hypoxia",
    `Main_diagnosis_code` %in% c("С71", "D33", "D32", "D43", "I72", "C71.2", "D33.3", "D33.1", "D33.0", "D43.0;G81.9", "D43.1", "D43.0", "D32.2", "D32.0", "C71.1", "I72.0", "I72.8") ~ "Tumor",
    `Main_diagnosis_code` %in% c("I67.2", "I67.1") ~ "Ischemic stroke",
    `Main_diagnosis_code` %in% c("I69.0", "I69.0;G91.0") ~ "Hemorrhagic stroke",
    `Main_diagnosis_code` %in% c("T06.8", "T91.3", "T98.1", "T82.8") ~ "TBI",
    `Main_diagnosis_code` %in% c("G93.1", "G82.4", "S15.0", "J17.1") ~ "Hypoxia",
    `Main_diagnosis_code` %in% c("D35.3;G04.2", "D35.2", "C79.3", "C71.8", "C71.6", "C71") ~ "Tumor",
    TRUE ~ NA
  ),
  `Main_diagnosis_code`
) 

data_e <- data_e %>% mutate(diseaseGroup = as.factor(diseaseGroup))

## Find ungrouped diagnosis
data_e %>%
  dplyr::select(`Main_diagnosis_code`, diseaseGroup) %>%
  filter(is.na(diseaseGroup))  %>%  unique() %>% arrange(.desc=T)

```

```{r fig.height=9, fig.width=10}
theme_cust <- theme(
    axis.text.x = element_text(size = 16, face = 'bold'),
    axis.text.y = element_text(size = 18),
    axis.title = element_text(size = 18),
    legend.title = element_text(size = 16),
    plot.title = element_text(hjust=0.5),
    legend.text = element_text(size = 16),
    panel.background = element_rect(fill = "white", color = 'black'),
    plot.background = element_rect(fill = "white"),
    ##panel.grid.major = element_line(color = "gray90"),
    ##panel.grid.minor = element_line(color = "lightgray"),
    strip.text = element_text(size = 16)
  )

data_e %>%
  group_by(diseaseGroup) %>%
  drop_na(diseaseGroup) %>% 
  summarise(count = n()) %>% 
  ggplot(aes(x = diseaseGroup, y = count)) +
  geom_bar(stat="identity", col = 'black', fill = 'white', linewidth=1) +
  geom_label(aes(label=count), col = 'black', size=10, label.size=1) +
  scale_x_discrete('Disease group') +
  scale_y_continuous('N patients') +
  theme_cust
```


## Patients status at FSCCRR delivery and demographics

```{r fig.width= 10, fig.height=8}

#SEX
ggplot(data_e %>% filter(!is.na(Sex))) +
  geom_bar(aes(y = diseaseGroup))+
  facet_wrap(~`Sex`, ncol = 1) +
  scale_y_discrete('Disease group') +
  scale_x_continuous('Number of patients') +
  theme_light() +
  theme_cust +
  theme(legend.position = "bottom")

#AGE
data_e %>% dplyr::filter (!(is.na(diseaseGroup)))%>%
ggplot() +
  geom_histogram (aes(`Age`, fill = diseaseGroup ), color = "grey80") + 
  facet_wrap(~diseaseGroup) +
  scale_y_discrete('Disease group') +
  scale_x_continuous('Age') +
  theme_light() +
  theme_base() +
  theme_cust +
  scale_fill_brewer(palette = "PiYG") +
  labs(fill = "Disease group")+
  theme(legend.position = "none") +
  labs (title = ' Disease group vs Age')


ggplot(data_e %>% filter(!is.na(diseaseGroup))) +
  geom_pointrange(mapping = aes(y =  diseaseGroup, x = Age, color = diseaseGroup ),
                  stat = "summary",
                  fun.min = function(z) {quantile(z,0.25)},
                  fun.max = function(z) {quantile(z,0.75)},
                  fun = median, 
                  size = 1) +
  facet_wrap(~`Sex`, ncol = 1) +
  scale_y_discrete('Disease group') +
  scale_x_continuous('Number of patients') +
  theme_base() +
  theme_cust +
  scale_color_brewer(palette = "Reds") 

##PNEUMONIA / INFECTION/ SEPSIS
inf <- data_e %>% dplyr::select (ID, Pneumonia, Infection, `Sepsis/septic_shock`, diseaseGroup ) %>% pivot_longer(!c(ID, diseaseGroup)) %>% filter (!(is.na(diseaseGroup )))
ggplot(inf) +
  geom_bar (aes(value, fill = name), position = "dodge") +
  facet_wrap( ~diseaseGroup ) +
  scale_y_continuous ('Number of patients') +
  scale_fill_brewer(palette = "Paired") +
  labs (title = 'Infection status vs Disease group') +
  theme_base () +
  theme_cust

```

```{r fig.width= 12, fig.height=10}
#CCI_1d
data_e %>% dplyr::filter (!(is.na(diseaseGroup))) %>% 
  filter (!(is.na(`CCI_1d`))) %>% 
  mutate(`CCI_1d` = fct_relevel(`CCI_1d`, 
            "0",'1','2','3','4','5', '6','7','8','9','10','11', '12','13',
            '14','15','16','17')) %>% 
  ggplot() +
  geom_bar (aes(`CCI_1d`, fill = diseaseGroup ), color = "grey80") +
  facet_wrap(~diseaseGroup) +
  scale_x_discrete('CCI 1 day') +
  scale_y_continuous('Number of patients') +
  theme_base() +
  theme_cust +
  theme(legend.position = "none") +
  labs (title = 'Charlson comorbidity score vs Disease group') +
   scale_fill_brewer(palette = "PiYG") 

#SOFA_1d
data_e %>% dplyr::filter (!(is.na(diseaseGroup))) %>% 
  filter (!(is.na(`SOFA_1d`))) %>% 
  mutate(`SOFA_1d` = fct_relevel(`SOFA_1d`, 
            "0",'1','2','3','4','5', '6','7','8','9','10','11', '12','13',
            '14','15','16','17')) %>% 
  ggplot() +
  geom_bar (aes(`SOFA_1d`, fill = diseaseGroup ), color = "grey80") +
  facet_wrap(~diseaseGroup) +
  scale_x_discrete('SOFA 1 day') +
  scale_y_continuous('Number of patients') +
  theme_base() +
  theme_cust +
  theme(legend.position = "none") +
  labs (title = 'SOFA score vs Disease group') +
   scale_fill_brewer(palette = "PiYG") 


   
```



## Dataset for analysis preparation

Prepare dataset for for PCA analysis and univariate regression, including 0-day status, demographic data, disease group and NLR. Also create a new additional categorial variable for bed-days at FSCCRR based on quantiles , called **BD_group**, for visualization purporses. 

```{r}

## calculate quantiles for bed-days  categorization 
`beddays at FSCCRR` <- as.numeric(data_e$`FSCCRR_beddays`)
hist (`beddays at FSCCRR`,  col = "gray78", breaks = 15)
BD <- as.numeric(data_e$`FSCCRR_beddays`)
res<-quantile(BD, probs = c(0,0.25,0.5,0.75,1), na.rm = T)
res

data_for_regres <- data_e %>% dplyr:: select (ID, Sex, Age, `anti-HCV`, `Hbs_AG`,  `LMCA`, Pneumonia, Infection, diseaseGroup, CGS, CCI_1d,`CIRS-G_sum_1d`, `CIRS-G_pathology_1d`, `CIRS-G_severity_1d`, FOUR_1d, SOFA_1d, `8ohdg_1d`, WBC_1d,  `NEU_#_1d`, `LYM_#_1d`,`PLT_1d`, `Lactate_1d`, Creatinine_1d, ALT_1d, AST_1d,  Albumin_1d, Alcaline_phosphatase_1d , Uric_acid_1d, Urea_1d, GGT_1d, Potassium_1d, Sodium_1d, Chloride_1d,  APTT_1d, INR_1d, Fibrinogen_1d, PT_1d, `Uroseptic_infection`, `Sepsis/septic_shock`, `Ventilation`, `Pneumonia_deterioration_d`, `Ventilation_d`, `ICU_FSCCRR_d`, `FSCCRR_beddays`, `Outcome`, `Pneumonia_at_FSCCRR`, `oneday_log`, `8ohdg 1d log`, `NLR_1d`) %>% 
  
  mutate(across(c(CGS, CCI_1d ,`CIRS-G_sum_1d`, `CIRS-G_pathology_1d`,`CIRS-G_severity_1d`,`FOUR_1d`, `SOFA_1d`, `WBC_1d`, `NEU_#_1d`, `LYM_#_1d`, `PLT_1d`, `Lactate_1d`, `Creatinine_1d` , `ALT_1d`, `AST_1d`, `Albumin_1d`, `Alcaline_phosphatase_1d`, `Uric_acid_1d`, `Urea_1d` , `GGT_1d`, `Potassium_1d`, `Sodium_1d`, `Chloride_1d`, `APTT_1d`, `INR_1d`,`Fibrinogen_1d`, PT_1d , `Pneumonia_deterioration_d`, `Ventilation_d`, `ICU_FSCCRR_d`, `FSCCRR_beddays`), as.numeric), 
         across(c(Outcome,`Ventilation`), factor)) %>%
  ## new variable - categorised `FSCCRR_beddays` by quantiles
  mutate (BD_group = case_when (`FSCCRR_beddays` < 30 ~ "0-30",
                                30 <= `FSCCRR_beddays` & `FSCCRR_beddays` < 46 ~ "30-46",
                                46 <= `FSCCRR_beddays` & `FSCCRR_beddays` < 67 ~ "46-67",
                                67 < `FSCCRR_beddays` ~ "67-401"))
  

 ## add corrected SNP data

hosp_poly_group_nlr  <- left_join(data_for_regres, hospital_polymorf_e[,1:11], by = "ID" ) %>%  
                              dplyr::select (!ID) %>% 
                              filter(!is.na(Ventilation))

 ##Achieve corrected  dataset with right names and correct variables types

 ##summary(hosp_poly_group_nlr)
 ##str(hosp_poly_group_nlr)
 ##colnames(hosp_poly_group_nlr)

hosp_poly_group_nlr %>% group_by(`BD_group`)  %>% tally ()

```

**Based on clinical data and analysis of cases count it was decided to chose only discharged patients (exclude '1'/dead cases) and evaluate ventilation demand as an outcome for analysis **

```{r}
 ## main dataset for discharged patients only
discharged <- hosp_poly_group_nlr %>% filter(Outcome == "0") %>%  dplyr::select (!c(Outcome, `Ventilation_d`)) %>%  filter(!is.na(Ventilation))

```

Outcome "ventilation demand" visualization 

```{r fig.width= 12, fig.height=10}

discharged  %>% dplyr::filter (!(is.na(diseaseGroup)))%>%
ggplot() +
  geom_boxplot (aes(Ventilation, Age, fill = Ventilation), color = "grey8") + 
  facet_wrap(~diseaseGroup) +
  scale_y_continuous('Age') +
 # scale_x_continuous('Age') +
  theme_light() +
  theme_base() +
  theme_cust +
  scale_fill_manual (values = c("Yes"= "#e34a33","No"= "#addd8e")) +
  labs(fill = "Disease group")+
  theme(legend.position = "none") +
  labs (title = 'Ventilation vs Disease group')

```

## Principal component analysis

Select numeric variables for day 0, demographic data, diagnosis group, and calculated NLR, DNA

```{r}

##str(hosp_poly_group_nlr) 

##check how many NA among chosen variables 

na_count <-sapply(discharged, function(y) sum(length(which(is.na(y)))))
na_count <- data.frame(na_count)

##delete variables with more than 60 NA's (more than 11%)

full_nums <- discharged %>% dplyr::select (!c(LMCA, `8ohdg_1d`,`Lactate_1d`,
                                                `Alcaline_phosphatase_1d`, `Uric_acid_1d`, `GGT_1d`, 
                                                `Chloride_1d`, `Pneumonia_deterioration_d`, `8ohdg 1d log`,`FSCCRR_beddays`)) %>% 
                            dplyr::select (BD_group, Ventilation, diseaseGroup, where(is.numeric)) %>% drop_na()
## got a tibble:302 × 29
```

Run PCA analysis for all numeric variables, chosen above

```{r}
hospital.full.pca <- prcomp(full_nums[,-(1:3)], 
                scale = T) # scaling
summary(hospital.full.pca)

contr_1a <- fviz_contrib(hospital.full.pca, choice = "var", axes = 1) 
contr_2a <- fviz_contrib(hospital.full.pca, choice = "var", axes = 2) 

fviz_eig (hospital.full.pca, addlabels = T, ylim = c (0,50))
pca_graph <- fviz_pca_var(hospital.full.pca, col.var = "contrib", gradient.cols = c("blue", "orange", "red"),
                          legend.title = "Contrib", pointsize = 10, pointshape = 40,
             labelsize = 5, repel = TRUE,  axes.linetype = "blank")
pca_graph
```
Started by classifying the variables into 3 groups using the k-means clustering algorithm. Next, we use the clusters returned by the k-means algorithm to color variables.


```{r}
set.seed(123)
var_pca_full <- get_pca_var(hospital.full.pca)
res.km <- kmeans(var_pca_full$coord, centers = 3, nstart = 25)
grp <- as.factor(res.km$cluster)
# Color variables by groups
fviz_pca_var(hospital.full.pca, col.var = grp, 
             palette =  c("#00AFBB", "#E7B800", "#FC4E07"),
             legend.title = "Cluster", pointsize = 10, pointshape = 40,
             labelsize = 5, repel = TRUE,  axes.linetype = "blank")
```

Biplot for Ventilation demand ( 0 = NO, 1 = YES)

```{r}
biplot0 <- ggbiplot(hospital.full.pca, 
         scale=0, alpha = 0.3, ellipse = TRUE, groups = full_nums$Ventilation) +
         labs(title = "PCA of yield contributing parameters")
biplot0
```

Biplot for categorised BED-DAYS at center ( "0-30", "30-46", "46-67", "67-401"  )

```{r}
biplot_BD <- ggbiplot(hospital.full.pca, 
         scale=0, alpha = 0.3, ellipse = TRUE, groups = full_nums$`BD_group`) +
         labs(title = "PCA of yield contributing parameters")
biplot_BD
```

Select less predictors, additionally exclude lab test results (except NLR) and components of CIRS-G composit score

```{r}
## Get less predictors, additionally exclude lab test results (except NLR) and components of CIRS-G composit score

unlist(colnames (full_nums))

small_pca <- discharged  %>% dplyr::select (Ventilation, BD_group, diseaseGroup, Age, CGS, CCI_1d,  `CIRS-G_sum_1d`, 
                                            `FOUR_1d`, `NLR_1d`, `ICU_FSCCRR_d`) %>% drop_na()


hospital.pca <- prcomp(small_pca[,-(1:3)], 
                scale = T) ## scaling
summary(hospital.pca)
## hospital.pca$rotation

## representation for described variance proportion for each component 
fviz_eig (hospital.pca, addlabels = T, ylim = c (0,50))

contrib_1 <- fviz_contrib(hospital.pca, choice = "var", axes = 1, labelsize = 5,) 
contrib_2 <- fviz_contrib(hospital.pca, choice = "var", axes = 2, labelsize = 5,) ## axes - component number, contributors to each component
contrib_1
contrib_2
pca_sm_graph <- fviz_pca_var(hospital.pca, col.var = "contrib",
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             pointsize = 3, pointshape = 21,
             labelsize = 5, repel = TRUE)
pca_sm_graph

```
The larger the value of the contribution, the more the variable contributes to the component. Contribution across all PC to highlight the most contributing variables for each dimension:

```{r}

var_pca <- get_pca_var(hospital.pca)

corrplot(var_pca$contrib, is.corr=FALSE)
```


Biplot for Ventilation demand ( 0 = NO, 1 = YES)

```{r}
biplot2 <- ggbiplot(hospital.pca, 
         scale=0, alpha = 0.3, ellipse = TRUE, groups = small_pca$Ventilation) +
         labs(title = "PCA - Ventilation biplot") +
         theme_base()
      
biplot2
```

Biplot for BED-DAYS at center ( "0-30", "30-46", "46-67", "67-401")

```{r}
biplot_BD_sm <- ggbiplot(hospital.pca, 
         scale=0, alpha = 0.3, ellipse = TRUE, groups = small_pca$`BD_group`) +
         labs(title = "PCA of yield contributing parameters") +
         labs(title = "PCA - Duration stay biplot") +
         theme_base()
biplot_BD_sm
```

Biplot for disease_group

```{r}
biplot_des_sm <- ggbiplot(hospital.pca, 
         scale=0, alpha = 0.3, ellipse = TRUE, groups = small_pca$`diseaseGroup`) +
         labs(title = "PCA - Disease groups biplot") +
        theme_base()
biplot_des_sm
```

3D visualization biplot for ventilation demand


```{r}

hospital.pca <- prcomp(small_pca[,-(1:3)], 
                scale = T )
##summary(hospital.pca)

components1 <- hospital.pca[["x"]]
     
components1 <- data.frame(components1[,1:3])
components1$PC2 <- -components1$PC2
components1$PC3 <- -components1$PC3
components1 = cbind(components1, small_pca$Ventilation)

##tot_explained_variance_ratio <- summary(hospital.pca)[["importance"]]['Proportion of Variance',]
##tot_explained_variance_ratio <- 100 * sum(tot_explained_variance_ratio)

tit = 'Total Explained Variance of first k=3 (out of 7) components = 77%'

fig <- plot_ly(components1, x = ~PC1, y = ~PC2, z = ~PC3, color = ~small_pca$Ventilation, colors = c('#EF553B','#00CC96') ) %>%
  add_markers(size = I(200))

fig <- fig %>%
  layout(
    title = tit,
    scene = list(bgcolor = "white")
)

fig

```

3D visualization biplot for groups by FSCCRR duration stay

```{r}
components1 <- cbind(components1, small_pca$BD_group)

tit <- 'Total Explained Variance of first k=3 (out of 7) components = 77%'

fig <- plot_ly(components1, x = ~PC1, y = ~PC2, z = ~PC3, color = ~small_pca$BD_group, colors = c('#EF553B','#00CC96', '#636EFA', '#EEB422')) %>%
  add_markers(size = I(200))


fig_BD <- fig %>%
  layout(
    title = tit,
    scene = list(bgcolor = "white")
)

fig_BD

```

3D visualization biplot for groups by disease

```{r}
components_d <- cbind(components1, small_pca$diseaseGroup)

tit <- 'Total Explained Variance of first k=3 (out of 7) components = 77%'

fig <- plot_ly(components_d, x = ~PC1, y = ~PC2, z = ~PC3, color = ~small_pca$diseaseGroup, colors = c('#EF553B','#00CC96', '#636EFA', '#EEB422')) %>%
  add_markers(size = I(200))

fig_dis <- fig %>%
  layout(
    title = tit,
    scene = list(bgcolor = "white")
)

fig_dis

```

Visualization of 4 components (grouped by Ventilation)

```{r eval = F}

hospital.pca <- prcomp(small_pca[,-(1:3)], 
                scale = T )
components2 <- hospital.pca[["x"]]
     
components2 <- data.frame(components2[,1:4])
components2$PC2 <- -components2$PC2
components2 = cbind(components2, small_pca$Ventilation)
colnames(components2)[5] = 'Ventilation'

tit = 'Total Explained Variance = 87'

axis = list(showline=FALSE,
            zeroline=FALSE,
            gridcolor='#ffff',
            ticklen=4)

fig2 <- components2 %>%
  plot_ly() %>%
  add_trace(
    type = 'splom',
    dimensions = list(
      list(label='PC1', values=~PC1),
      list(label='PC2', values=~PC2),
      list(label='PC3', values=~PC3),
      list(label='PC4', values=~PC4)
    ),
    color=~Ventilation,
    marker = list(
      size = 7
    )
  ) %>% style(diagonal = list(visible = FALSE)) %>%
  layout(
    title= tit,
    hovermode='closest',
    dragmode= 'select',
    plot_bgcolor='rgba(240,240,240, 0.95)',
    xaxis=list(domain=NULL, showline=F, zeroline=F, gridcolor='#ffff', ticklen=4),
    yaxis=list(domain=NULL, showline=F, zeroline=F, gridcolor='#ffff', ticklen=4),
    xaxis2=axis,
    xaxis3=axis,
    xaxis4=axis,
    yaxis2=axis,
    yaxis3=axis,
    yaxis4=axis
  )
options(warn=-1)
fig2
```


## Univariate regression for all variables (at time point 0)

 Run regression for all variables of interest

```{r}

x <- names(discharged[,-39])
formulas <- unlist(lapply(39, function(n) combn(x, 1, FUN=function(row) paste0("Ventilation ~ ", paste0("`",row, "`",collapse = "+")))))
formulas  ## combine each variable into formula ("outcome ~ variable of interest")
```

```{r warning=FALSE}

##regression coefficients
tmp1 <- bind_rows(lapply(formulas, function(frml) {
 a <- tidy(glm(frml, data = discharged, family = binomial), exponentiate = TRUE, conf.int = TRUE) %>%
   mutate(across(where(is.numeric), round, digits = 2))
 a$frml = frml
 return(a)
}))

tmp1 <- tmp1 %>% mutate (`sign_.05` = ifelse (`p.value` >= 0.05, "no", "yes" ))

signif_estim_univar <- as_data_frame(tmp1) %>% filter (`p.value` <= 0.05)

write.xlsx(signif_estim_univar, 'signif_estim_univar_glm.xlsx')
##regression results R2, AIC, BIC
##tmp2 = bind_rows(lapply(out, function(frml) {
## a = glance(lm(frml, data=hosp_poly_group_nlr))
## a$frml = frml
## return(a)}))
##signif_results_univar<- as_data_frame(tmp2)
##write.xlsx(signif_estim_univar, 'signif_results_univar_glm.xlsx')

```

```{r warning = F, eval = F}
models <- formulas %>%               ##add all formulas
  
  ##iterate through each univariate formula
  map(                               
    .f = ~glm(                       ## pass the formulas one-by-one to glm()
      formula = as.formula(.x),      ## within glm(), the string formula is .x
      family = "binomial",           ## specify type of glm (logistic)
      data = discharged)) %>%        ## dataset
  
  ## tidy up each of the glm regression outputs from above
  map(
    .f = ~tidy(
      .x, 
      exponentiate = TRUE,           ## exponentiate 
      conf.int = TRUE)) %>%          ## return confidence intervals
  
  ## collapse the list of regression outputs in to one data frame
  bind_rows() %>% 
  ## round all numeric columns
  mutate(across(where(is.numeric), round, digits = 2)) 
```

```{r}

univ_tab <- discharged %>% 
  dplyr::select(all_of(x), `Ventilation`) %>% ## select variables of interest

  tbl_uvregression(                         ## produce univariate table
    method = glm,                           ## define regression want to run (generalised linear model)
    y = Ventilation,                        ## define outcome variable
    method.args = list(family = binomial),  ## define what type of glm want to run (logistic)
    exponentiate = TRUE                     ## exponentiate to produce odds ratios (rather than log odds)
  ) %>% 
bold_p() %>% 
add_q() %>%
filter_p(t = 0.05) 

## view univariate results table 
univ_tab
```

Graphical representation of univariant regression (color designate statistical significance)

```{r fig.width=20, fig.height=10}

#tmp1[-c(9,81),]
tmp1%>% 
  
  #set order of levels to appear along y-axis
  mutate(`sign_.05` = fct_relevel(
    `sign_.05`, "yes", "no")) %>%
  
  # remove "intercept" row from plot
  filter(term != "(Intercept)") %>% 
  
   # remove "UNSIGNIFICANT" TERMS from plot
  filter (`sign_.05`!= "no") %>% 
  
  ## plot with variable on the y axis and estimate (OR) on the x axis
  ggplot(aes(x = estimate, y = term)) +
  
  ## show the estimate as a point
  geom_point(aes(color = `sign_.05`)) + 
  
  ## add in an error bar for the confidence intervals
  geom_errorbar(aes(xmin = conf.low, xmax = conf.high), width = 0.5) + 
  
  ## show where OR = 1 is for reference as a dashed line
  geom_vline(xintercept = 1, linetype = "dotted") +
  scale_x_continuous(limits = c(-0.5, 25)) +  # cut off extreme CI range
  theme_base() +
  theme_cust +
  labs(title = "Result of univariate regression")
  
```


### Multivariate regression for Аssociation TLR9 and log(extracellular DNA)


```{r}

fit_tlr9 <- glm (`Ventilation` ~  `TLR9_rs352162` * `oneday_log` + `TLR9_rs187084` * `oneday_log`+ CGS + `CCI_1d`, data = discharged, family = binomial)
summary(fit_tlr9)

fit_tlr9_gr <- glm (`Ventilation` ~  `TLR9_rs352162` * `oneday_log` + `TLR9_rs187084` * `oneday_log`+ CGS + `CCI_1d` + diseaseGroup, data = discharged, family = binomial)
summary(fit_tlr9_gr)


#discharged$TLR9_rs352162
#discharged$TLR9_rs187084
#discharged$AQP4_rs1058427
```